{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ZPFmKWEDygR6"
      },
      "source": [
        "![Cover](https://raw.githubusercontent.com/crunchdao/quickstarters/refs/heads/master/competitions/broad-1/assets/cover.png)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "F8PyJPSkxoRo"
      },
      "source": [
        "## Autoimmune Disease Machine Learning Crunch #1\n",
        "\n",
        "## Quickstarter: Resnet50 + Ridge\n",
        "\n",
        "This is an example submission and can be used as a starting point, but should not be seen as the only way to approach the problem. However, this approach can be improved in many ways."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ovnPAKW52GYm"
      },
      "source": [
        "This notebook addresses the task of **predicting gene expression** in spatial transcriptomics data of colon tissue using H&E pathology images. The goal is to predict the expression of **460 genes** in tissue patches based on H&E images and corresponding spatial transcriptomics data.\n",
        "\n",
        "The solution follows a two-step process:\n",
        "\n",
        "- **ResNet50 Embedding**: We use the pre-trained ResNet50 model, trained on ImageNet, to extract features from the H&E images. The model transforms the high-dimensional image data into a lower-dimensional numerical representation (embedding vector).\n",
        "- **Ridge Regression**: After obtaining the embedding vectors, ridge regression is applied to predict the gene expression profiles for each nucleus/cell.\n",
        "\n",
        "> For CrunchDAO Challenge submissions, where internet access is restricted, you will need to **pre-download** the model file `pytorch_model.bin` from its [Resnet50 Hugging Face](https://huggingface.co/timm/resnet50.tv_in1k/tree/main) card (link: https://huggingface.co/timm/resnet50.tv_in1k/tree/main) and submit it along with the notebook [(see how submit below)](#scrollTo=kppG1inFIopT&line=8&uniqifier=1).\n",
        "\n",
        "This code is adapted from the GitHub repository [HEST-library](https://github.com/mahmoodlab/hest) (link: https://github.com/mahmoodlab/hest). Dive into the repository to explore advanced pre-trained embedding models, experiment with alternative regression techniques or tap into external datasets to push the boundaries of performance!\n",
        "\n",
        "For exploring other pre-trained models, you can also visit:\n",
        "\n",
        "- Hugging Face Timm: https://huggingface.co/timm\n",
        "- PyTorch Vision Models: https://pytorch.org/vision/stable/models.html\n",
        "\n",
        "Note: This notebook is designed to run seamlessly on a **CPU** environment. (25 min for train step with default values - Mean Squared Error of ~0.10 and Pearson correlation coefficient of ~0.15 on unseen H&E data) -- (1h for submission)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "SWQHlrgiNekV"
      },
      "source": [
        "`Predicting spatial transcriptomics data from an H&E image (Crunch 1)`\n",
        "\n",
        "![crunch_1_challenge](https://raw.githubusercontent.com/crunchdao/competitions/refs/heads/master/competitions/broad-1/quickstarters/resnet50-plus-ridge/images/crunch_1_challenge.png)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "9UjSFajKnBnq"
      },
      "source": [
        "## Setup"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "iKVOi6J1loBn"
      },
      "source": [
        "You need a **token** to load the dataset of the challenge.\n",
        "\n",
        "Get a new token: https://hub.crunchdao.com/competitions/broad-1/submit/via/notebook"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "p9pR4gsI9VE2",
        "outputId": "a639bc61-facf-447e-e23e-f30a0536ef02"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting crunch-cli\n",
            "  Downloading crunch_cli-5.7.5-py3-none-any.whl.metadata (3.1 kB)\n",
            "Requirement already satisfied: click in /usr/local/lib/python3.10/dist-packages (from crunch-cli) (8.1.7)\n",
            "Collecting coloredlogs (from crunch-cli)\n",
            "  Downloading coloredlogs-15.0.1-py2.py3-none-any.whl.metadata (12 kB)\n",
            "Collecting dataclasses_json (from crunch-cli)\n",
            "  Downloading dataclasses_json-0.6.7-py3-none-any.whl.metadata (25 kB)\n",
            "Requirement already satisfied: importlib_metadata in /usr/local/lib/python3.10/dist-packages (from crunch-cli) (8.5.0)\n",
            "Collecting inquirer (from crunch-cli)\n",
            "  Downloading inquirer-3.4.0-py3-none-any.whl.metadata (6.8 kB)\n",
            "Requirement already satisfied: joblib in /usr/local/lib/python3.10/dist-packages (from crunch-cli) (1.4.2)\n",
            "Requirement already satisfied: networkx in /usr/local/lib/python3.10/dist-packages (from crunch-cli) (3.4.2)\n",
            "Requirement already satisfied: packaging>=24.2 in /usr/local/lib/python3.10/dist-packages (from crunch-cli) (24.2)\n",
            "Requirement already satisfied: pandas in /usr/local/lib/python3.10/dist-packages (from crunch-cli) (2.2.2)\n",
            "Requirement already satisfied: psutil in /usr/local/lib/python3.10/dist-packages (from crunch-cli) (5.9.5)\n",
            "Requirement already satisfied: pyarrow in /usr/local/lib/python3.10/dist-packages (from crunch-cli) (17.0.0)\n",
            "Requirement already satisfied: pytest in /usr/local/lib/python3.10/dist-packages (from crunch-cli) (8.3.4)\n",
            "Collecting python-dotenv (from crunch-cli)\n",
            "  Downloading python_dotenv-1.0.1-py3-none-any.whl.metadata (23 kB)\n",
            "Requirement already satisfied: pyyaml in /usr/local/lib/python3.10/dist-packages (from crunch-cli) (6.0.2)\n",
            "Collecting redbaron (from crunch-cli)\n",
            "  Downloading redbaron-0.9.2-py2.py3-none-any.whl.metadata (15 kB)\n",
            "Requirement already satisfied: requests in /usr/local/lib/python3.10/dist-packages (from crunch-cli) (2.32.3)\n",
            "Requirement already satisfied: requests-toolbelt in /usr/local/lib/python3.10/dist-packages (from crunch-cli) (1.0.0)\n",
            "Requirement already satisfied: requirements-parser in /usr/local/lib/python3.10/dist-packages (from crunch-cli) (0.9.0)\n",
            "Requirement already satisfied: scikit-learn in /usr/local/lib/python3.10/dist-packages (from crunch-cli) (1.6.0)\n",
            "Requirement already satisfied: scipy in /usr/local/lib/python3.10/dist-packages (from crunch-cli) (1.13.1)\n",
            "Requirement already satisfied: tqdm in /usr/local/lib/python3.10/dist-packages (from crunch-cli) (4.67.1)\n",
            "Collecting spatialdata-plot (from crunch-cli)\n",
            "  Downloading spatialdata_plot-0.2.8-py3-none-any.whl.metadata (6.8 kB)\n",
            "Collecting humanfriendly>=9.1 (from coloredlogs->crunch-cli)\n",
            "  Downloading humanfriendly-10.0-py2.py3-none-any.whl.metadata (9.2 kB)\n",
            "Collecting marshmallow<4.0.0,>=3.18.0 (from dataclasses_json->crunch-cli)\n",
            "  Downloading marshmallow-3.23.2-py3-none-any.whl.metadata (7.1 kB)\n",
            "Collecting typing-inspect<1,>=0.4.0 (from dataclasses_json->crunch-cli)\n",
            "  Downloading typing_inspect-0.9.0-py3-none-any.whl.metadata (1.5 kB)\n",
            "Requirement already satisfied: zipp>=3.20 in /usr/local/lib/python3.10/dist-packages (from importlib_metadata->crunch-cli) (3.21.0)\n",
            "Collecting blessed>=1.19.0 (from inquirer->crunch-cli)\n",
            "  Downloading blessed-1.20.0-py2.py3-none-any.whl.metadata (13 kB)\n",
            "Collecting editor>=1.6.0 (from inquirer->crunch-cli)\n",
            "  Downloading editor-1.6.6-py3-none-any.whl.metadata (2.3 kB)\n",
            "Collecting readchar>=4.2.0 (from inquirer->crunch-cli)\n",
            "  Downloading readchar-4.2.1-py3-none-any.whl.metadata (7.5 kB)\n",
            "Requirement already satisfied: numpy>=1.22.4 in /usr/local/lib/python3.10/dist-packages (from pandas->crunch-cli) (1.26.4)\n",
            "Requirement already satisfied: python-dateutil>=2.8.2 in /usr/local/lib/python3.10/dist-packages (from pandas->crunch-cli) (2.8.2)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.10/dist-packages (from pandas->crunch-cli) (2024.2)\n",
            "Requirement already satisfied: tzdata>=2022.7 in /usr/local/lib/python3.10/dist-packages (from pandas->crunch-cli) (2024.2)\n",
            "Requirement already satisfied: exceptiongroup>=1.0.0rc8 in /usr/local/lib/python3.10/dist-packages (from pytest->crunch-cli) (1.2.2)\n",
            "Requirement already satisfied: iniconfig in /usr/local/lib/python3.10/dist-packages (from pytest->crunch-cli) (2.0.0)\n",
            "Requirement already satisfied: pluggy<2,>=1.5 in /usr/local/lib/python3.10/dist-packages (from pytest->crunch-cli) (1.5.0)\n",
            "Requirement already satisfied: tomli>=1 in /usr/local/lib/python3.10/dist-packages (from pytest->crunch-cli) (2.2.1)\n",
            "Collecting baron>=0.7 (from redbaron->crunch-cli)\n",
            "  Downloading baron-0.10.1-py2.py3-none-any.whl.metadata (16 kB)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.10/dist-packages (from requests->crunch-cli) (3.4.0)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.10/dist-packages (from requests->crunch-cli) (3.10)\n",
            "Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.10/dist-packages (from requests->crunch-cli) (2.2.3)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.10/dist-packages (from requests->crunch-cli) (2024.12.14)\n",
            "Requirement already satisfied: types-setuptools>=69.1.0 in /usr/local/lib/python3.10/dist-packages (from requirements-parser->crunch-cli) (75.6.0.20241126)\n",
            "Requirement already satisfied: threadpoolctl>=3.1.0 in /usr/local/lib/python3.10/dist-packages (from scikit-learn->crunch-cli) (3.5.0)\n",
            "Requirement already satisfied: matplotlib in /usr/local/lib/python3.10/dist-packages (from spatialdata-plot->crunch-cli) (3.8.0)\n",
            "Collecting matplotlib-scalebar (from spatialdata-plot->crunch-cli)\n",
            "  Downloading matplotlib_scalebar-0.8.1-py2.py3-none-any.whl.metadata (13 kB)\n",
            "Collecting scanpy (from spatialdata-plot->crunch-cli)\n",
            "  Downloading scanpy-1.10.4-py3-none-any.whl.metadata (9.3 kB)\n",
            "Collecting spatialdata>=0.2.6 (from spatialdata-plot->crunch-cli)\n",
            "  Downloading spatialdata-0.2.6-py3-none-any.whl.metadata (9.6 kB)\n",
            "Collecting rply (from baron>=0.7->redbaron->crunch-cli)\n",
            "  Downloading rply-0.7.8-py2.py3-none-any.whl.metadata (4.2 kB)\n",
            "Requirement already satisfied: wcwidth>=0.1.4 in /usr/local/lib/python3.10/dist-packages (from blessed>=1.19.0->inquirer->crunch-cli) (0.2.13)\n",
            "Requirement already satisfied: six>=1.9.0 in /usr/local/lib/python3.10/dist-packages (from blessed>=1.19.0->inquirer->crunch-cli) (1.17.0)\n",
            "Collecting runs (from editor>=1.6.0->inquirer->crunch-cli)\n",
            "  Downloading runs-1.2.2-py3-none-any.whl.metadata (10 kB)\n",
            "Collecting xmod (from editor>=1.6.0->inquirer->crunch-cli)\n",
            "  Downloading xmod-1.8.1-py3-none-any.whl.metadata (1.8 kB)\n",
            "Collecting anndata>=0.9.1 (from spatialdata>=0.2.6->spatialdata-plot->crunch-cli)\n",
            "  Downloading anndata-0.11.1-py3-none-any.whl.metadata (8.2 kB)\n",
            "Collecting dask-image (from spatialdata>=0.2.6->spatialdata-plot->crunch-cli)\n",
            "  Downloading dask_image-2024.5.3-py3-none-any.whl.metadata (2.8 kB)\n",
            "Requirement already satisfied: dask>=2024.4.1 in /usr/local/lib/python3.10/dist-packages (from spatialdata>=0.2.6->spatialdata-plot->crunch-cli) (2024.10.0)\n",
            "Collecting fsspec<=2023.6 (from spatialdata>=0.2.6->spatialdata-plot->crunch-cli)\n",
            "  Downloading fsspec-2023.6.0-py3-none-any.whl.metadata (6.7 kB)\n",
            "Requirement already satisfied: geopandas>=0.14 in /usr/local/lib/python3.10/dist-packages (from spatialdata>=0.2.6->spatialdata-plot->crunch-cli) (1.0.1)\n",
            "Collecting multiscale-spatial-image>=2.0.2 (from spatialdata>=0.2.6->spatialdata-plot->crunch-cli)\n",
            "  Downloading multiscale_spatial_image-2.0.2-py3-none-any.whl.metadata (25 kB)\n",
            "Requirement already satisfied: numba in /usr/local/lib/python3.10/dist-packages (from spatialdata>=0.2.6->spatialdata-plot->crunch-cli) (0.60.0)\n",
            "Collecting ome-zarr>=0.8.4 (from spatialdata>=0.2.6->spatialdata-plot->crunch-cli)\n",
            "  Downloading ome_zarr-0.10.2-py3-none-any.whl.metadata (2.9 kB)\n",
            "Requirement already satisfied: pooch in /usr/local/lib/python3.10/dist-packages (from spatialdata>=0.2.6->spatialdata-plot->crunch-cli) (1.8.2)\n",
            "Requirement already satisfied: rich in /usr/local/lib/python3.10/dist-packages (from spatialdata>=0.2.6->spatialdata-plot->crunch-cli) (13.9.4)\n",
            "Requirement already satisfied: scikit-image in /usr/local/lib/python3.10/dist-packages (from spatialdata>=0.2.6->spatialdata-plot->crunch-cli) (0.25.0)\n",
            "Requirement already satisfied: setuptools in /usr/local/lib/python3.10/dist-packages (from spatialdata>=0.2.6->spatialdata-plot->crunch-cli) (75.1.0)\n",
            "Requirement already satisfied: shapely>=2.0.1 in /usr/local/lib/python3.10/dist-packages (from spatialdata>=0.2.6->spatialdata-plot->crunch-cli) (2.0.6)\n",
            "Collecting spatial-image>=1.1.0 (from spatialdata>=0.2.6->spatialdata-plot->crunch-cli)\n",
            "  Downloading spatial_image-1.1.0-py3-none-any.whl.metadata (5.5 kB)\n",
            "Requirement already satisfied: typing-extensions>=4.8.0 in /usr/local/lib/python3.10/dist-packages (from spatialdata>=0.2.6->spatialdata-plot->crunch-cli) (4.12.2)\n",
            "Collecting xarray-schema (from spatialdata>=0.2.6->spatialdata-plot->crunch-cli)\n",
            "  Downloading xarray_schema-0.0.3-py3-none-any.whl.metadata (4.3 kB)\n",
            "Collecting xarray-spatial>=0.3.5 (from spatialdata>=0.2.6->spatialdata-plot->crunch-cli)\n",
            "  Downloading xarray_spatial-0.4.0-py3-none-any.whl.metadata (16 kB)\n",
            "Requirement already satisfied: xarray>=2024.10.0 in /usr/local/lib/python3.10/dist-packages (from spatialdata>=0.2.6->spatialdata-plot->crunch-cli) (2024.11.0)\n",
            "Collecting zarr<3 (from spatialdata>=0.2.6->spatialdata-plot->crunch-cli)\n",
            "  Downloading zarr-2.18.3-py3-none-any.whl.metadata (5.7 kB)\n",
            "Collecting mypy-extensions>=0.3.0 (from typing-inspect<1,>=0.4.0->dataclasses_json->crunch-cli)\n",
            "  Downloading mypy_extensions-1.0.0-py3-none-any.whl.metadata (1.1 kB)\n",
            "Requirement already satisfied: contourpy>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib->spatialdata-plot->crunch-cli) (1.3.1)\n",
            "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.10/dist-packages (from matplotlib->spatialdata-plot->crunch-cli) (0.12.1)\n",
            "Requirement already satisfied: fonttools>=4.22.0 in /usr/local/lib/python3.10/dist-packages (from matplotlib->spatialdata-plot->crunch-cli) (4.55.3)\n",
            "Requirement already satisfied: kiwisolver>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib->spatialdata-plot->crunch-cli) (1.4.7)\n",
            "Requirement already satisfied: pillow>=6.2.0 in /usr/local/lib/python3.10/dist-packages (from matplotlib->spatialdata-plot->crunch-cli) (11.0.0)\n",
            "Requirement already satisfied: pyparsing>=2.3.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib->spatialdata-plot->crunch-cli) (3.2.0)\n",
            "Requirement already satisfied: h5py>=3.6 in /usr/local/lib/python3.10/dist-packages (from scanpy->spatialdata-plot->crunch-cli) (3.12.1)\n",
            "Collecting legacy-api-wrap>=1.4 (from scanpy->spatialdata-plot->crunch-cli)\n",
            "  Downloading legacy_api_wrap-1.4.1-py3-none-any.whl.metadata (2.1 kB)\n",
            "Requirement already satisfied: natsort in /usr/local/lib/python3.10/dist-packages (from scanpy->spatialdata-plot->crunch-cli) (8.4.0)\n",
            "Requirement already satisfied: patsy!=1.0.0 in /usr/local/lib/python3.10/dist-packages (from scanpy->spatialdata-plot->crunch-cli) (1.0.1)\n",
            "Collecting pynndescent>=0.5 (from scanpy->spatialdata-plot->crunch-cli)\n",
            "  Downloading pynndescent-0.5.13-py3-none-any.whl.metadata (6.8 kB)\n",
            "Requirement already satisfied: seaborn>=0.13 in /usr/local/lib/python3.10/dist-packages (from scanpy->spatialdata-plot->crunch-cli) (0.13.2)\n",
            "Collecting session-info (from scanpy->spatialdata-plot->crunch-cli)\n",
            "  Downloading session_info-1.0.0.tar.gz (24 kB)\n",
            "  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "Requirement already satisfied: statsmodels>=0.13 in /usr/local/lib/python3.10/dist-packages (from scanpy->spatialdata-plot->crunch-cli) (0.14.4)\n",
            "Collecting umap-learn!=0.5.0,>=0.5 (from scanpy->spatialdata-plot->crunch-cli)\n",
            "  Downloading umap_learn-0.5.7-py3-none-any.whl.metadata (21 kB)\n",
            "Collecting array-api-compat!=1.5,>1.4 (from anndata>=0.9.1->spatialdata>=0.2.6->spatialdata-plot->crunch-cli)\n",
            "  Downloading array_api_compat-1.10.0-py3-none-any.whl.metadata (1.6 kB)\n",
            "Requirement already satisfied: cloudpickle>=3.0.0 in /usr/local/lib/python3.10/dist-packages (from dask>=2024.4.1->spatialdata>=0.2.6->spatialdata-plot->crunch-cli) (3.1.0)\n",
            "Requirement already satisfied: partd>=1.4.0 in /usr/local/lib/python3.10/dist-packages (from dask>=2024.4.1->spatialdata>=0.2.6->spatialdata-plot->crunch-cli) (1.4.2)\n",
            "Requirement already satisfied: toolz>=0.10.0 in /usr/local/lib/python3.10/dist-packages (from dask>=2024.4.1->spatialdata>=0.2.6->spatialdata-plot->crunch-cli) (0.12.1)\n",
            "Requirement already satisfied: pyogrio>=0.7.2 in /usr/local/lib/python3.10/dist-packages (from geopandas>=0.14->spatialdata>=0.2.6->spatialdata-plot->crunch-cli) (0.10.0)\n",
            "Requirement already satisfied: pyproj>=3.3.0 in /usr/local/lib/python3.10/dist-packages (from geopandas>=0.14->spatialdata>=0.2.6->spatialdata-plot->crunch-cli) (3.7.0)\n",
            "Requirement already satisfied: llvmlite<0.44,>=0.43.0dev0 in /usr/local/lib/python3.10/dist-packages (from numba->spatialdata>=0.2.6->spatialdata-plot->crunch-cli) (0.43.0)\n",
            "Collecting distributed (from ome-zarr>=0.8.4->spatialdata>=0.2.6->spatialdata-plot->crunch-cli)\n",
            "  Downloading distributed-2024.12.1-py3-none-any.whl.metadata (3.3 kB)\n",
            "Requirement already satisfied: aiohttp<4 in /usr/local/lib/python3.10/dist-packages (from ome-zarr>=0.8.4->spatialdata>=0.2.6->spatialdata-plot->crunch-cli) (3.11.10)\n",
            "Collecting xarray-dataclasses>=1.1.0 (from spatial-image>=1.1.0->spatialdata>=0.2.6->spatialdata-plot->crunch-cli)\n",
            "  Downloading xarray_dataclasses-1.9.1-py3-none-any.whl.metadata (10 kB)\n",
            "Collecting datashader>=0.15.0 (from xarray-spatial>=0.3.5->spatialdata>=0.2.6->spatialdata-plot->crunch-cli)\n",
            "  Downloading datashader-0.16.3-py2.py3-none-any.whl.metadata (12 kB)\n",
            "Collecting asciitree (from zarr<3->spatialdata>=0.2.6->spatialdata-plot->crunch-cli)\n",
            "  Downloading asciitree-0.3.3.tar.gz (4.0 kB)\n",
            "  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "Collecting numcodecs>=0.10.0 (from zarr<3->spatialdata>=0.2.6->spatialdata-plot->crunch-cli)\n",
            "  Downloading numcodecs-0.13.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (2.9 kB)\n",
            "Collecting fasteners (from zarr<3->spatialdata>=0.2.6->spatialdata-plot->crunch-cli)\n",
            "  Downloading fasteners-0.19-py3-none-any.whl.metadata (4.9 kB)\n",
            "Collecting pims>=0.4.1 (from dask-image->spatialdata>=0.2.6->spatialdata-plot->crunch-cli)\n",
            "  Downloading pims-0.7.tar.gz (87 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m87.8/87.8 kB\u001b[0m \u001b[31m5.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "Requirement already satisfied: tifffile>=2018.10.18 in /usr/local/lib/python3.10/dist-packages (from dask-image->spatialdata>=0.2.6->spatialdata-plot->crunch-cli) (2024.12.12)\n",
            "Requirement already satisfied: platformdirs>=2.5.0 in /usr/local/lib/python3.10/dist-packages (from pooch->spatialdata>=0.2.6->spatialdata-plot->crunch-cli) (4.3.6)\n",
            "Requirement already satisfied: markdown-it-py>=2.2.0 in /usr/local/lib/python3.10/dist-packages (from rich->spatialdata>=0.2.6->spatialdata-plot->crunch-cli) (3.0.0)\n",
            "Requirement already satisfied: pygments<3.0.0,>=2.13.0 in /usr/local/lib/python3.10/dist-packages (from rich->spatialdata>=0.2.6->spatialdata-plot->crunch-cli) (2.18.0)\n",
            "Collecting appdirs (from rply->baron>=0.7->redbaron->crunch-cli)\n",
            "  Downloading appdirs-1.4.4-py2.py3-none-any.whl.metadata (9.0 kB)\n",
            "Requirement already satisfied: imageio!=2.35.0,>=2.33 in /usr/local/lib/python3.10/dist-packages (from scikit-image->spatialdata>=0.2.6->spatialdata-plot->crunch-cli) (2.36.1)\n",
            "Requirement already satisfied: lazy-loader>=0.4 in /usr/local/lib/python3.10/dist-packages (from scikit-image->spatialdata>=0.2.6->spatialdata-plot->crunch-cli) (0.4)\n",
            "Collecting stdlib_list (from session-info->scanpy->spatialdata-plot->crunch-cli)\n",
            "  Downloading stdlib_list-0.11.0-py3-none-any.whl.metadata (3.3 kB)\n",
            "Requirement already satisfied: aiohappyeyeballs>=2.3.0 in /usr/local/lib/python3.10/dist-packages (from aiohttp<4->ome-zarr>=0.8.4->spatialdata>=0.2.6->spatialdata-plot->crunch-cli) (2.4.4)\n",
            "Requirement already satisfied: aiosignal>=1.1.2 in /usr/local/lib/python3.10/dist-packages (from aiohttp<4->ome-zarr>=0.8.4->spatialdata>=0.2.6->spatialdata-plot->crunch-cli) (1.3.2)\n",
            "Requirement already satisfied: async-timeout<6.0,>=4.0 in /usr/local/lib/python3.10/dist-packages (from aiohttp<4->ome-zarr>=0.8.4->spatialdata>=0.2.6->spatialdata-plot->crunch-cli) (4.0.3)\n",
            "Requirement already satisfied: attrs>=17.3.0 in /usr/local/lib/python3.10/dist-packages (from aiohttp<4->ome-zarr>=0.8.4->spatialdata>=0.2.6->spatialdata-plot->crunch-cli) (24.3.0)\n",
            "Requirement already satisfied: frozenlist>=1.1.1 in /usr/local/lib/python3.10/dist-packages (from aiohttp<4->ome-zarr>=0.8.4->spatialdata>=0.2.6->spatialdata-plot->crunch-cli) (1.5.0)\n",
            "Requirement already satisfied: multidict<7.0,>=4.5 in /usr/local/lib/python3.10/dist-packages (from aiohttp<4->ome-zarr>=0.8.4->spatialdata>=0.2.6->spatialdata-plot->crunch-cli) (6.1.0)\n",
            "Requirement already satisfied: propcache>=0.2.0 in /usr/local/lib/python3.10/dist-packages (from aiohttp<4->ome-zarr>=0.8.4->spatialdata>=0.2.6->spatialdata-plot->crunch-cli) (0.2.1)\n",
            "Requirement already satisfied: yarl<2.0,>=1.17.0 in /usr/local/lib/python3.10/dist-packages (from aiohttp<4->ome-zarr>=0.8.4->spatialdata>=0.2.6->spatialdata-plot->crunch-cli) (1.18.3)\n",
            "Collecting dask-expr<1.2,>=1.1 (from dask[array,dataframe]>=2024.4.1->dask-image->spatialdata>=0.2.6->spatialdata-plot->crunch-cli)\n",
            "  Downloading dask_expr-1.1.21-py3-none-any.whl.metadata (2.6 kB)\n",
            "Requirement already satisfied: colorcet in /usr/local/lib/python3.10/dist-packages (from datashader>=0.15.0->xarray-spatial>=0.3.5->spatialdata>=0.2.6->spatialdata-plot->crunch-cli) (3.1.0)\n",
            "Requirement already satisfied: multipledispatch in /usr/local/lib/python3.10/dist-packages (from datashader>=0.15.0->xarray-spatial>=0.3.5->spatialdata>=0.2.6->spatialdata-plot->crunch-cli) (1.0.0)\n",
            "Requirement already satisfied: param in /usr/local/lib/python3.10/dist-packages (from datashader>=0.15.0->xarray-spatial>=0.3.5->spatialdata>=0.2.6->spatialdata-plot->crunch-cli) (2.2.0)\n",
            "Collecting pyct (from datashader>=0.15.0->xarray-spatial>=0.3.5->spatialdata>=0.2.6->spatialdata-plot->crunch-cli)\n",
            "  Downloading pyct-0.5.0-py2.py3-none-any.whl.metadata (7.4 kB)\n",
            "Collecting s3fs (from fsspec[s3]!=2021.07.0,>=0.8->ome-zarr>=0.8.4->spatialdata>=0.2.6->spatialdata-plot->crunch-cli)\n",
            "  Downloading s3fs-2024.12.0-py3-none-any.whl.metadata (1.6 kB)\n",
            "Requirement already satisfied: mdurl~=0.1 in /usr/local/lib/python3.10/dist-packages (from markdown-it-py>=2.2.0->rich->spatialdata>=0.2.6->spatialdata-plot->crunch-cli) (0.1.2)\n",
            "Requirement already satisfied: locket in /usr/local/lib/python3.10/dist-packages (from partd>=1.4.0->dask>=2024.4.1->spatialdata>=0.2.6->spatialdata-plot->crunch-cli) (1.0.0)\n",
            "Collecting slicerator>=0.9.8 (from pims>=0.4.1->dask-image->spatialdata>=0.2.6->spatialdata-plot->crunch-cli)\n",
            "  Downloading slicerator-1.1.0-py3-none-any.whl.metadata (1.9 kB)\n",
            "INFO: pip is looking at multiple versions of distributed to determine which version is compatible with other requirements. This could take a while.\n",
            "Collecting distributed (from ome-zarr>=0.8.4->spatialdata>=0.2.6->spatialdata-plot->crunch-cli)\n",
            "  Downloading distributed-2024.12.0-py3-none-any.whl.metadata (3.3 kB)\n",
            "  Downloading distributed-2024.11.2-py3-none-any.whl.metadata (3.3 kB)\n",
            "  Downloading distributed-2024.11.1-py3-none-any.whl.metadata (3.3 kB)\n",
            "  Downloading distributed-2024.11.0-py3-none-any.whl.metadata (3.3 kB)\n",
            "  Downloading distributed-2024.10.0-py3-none-any.whl.metadata (3.3 kB)\n",
            "Requirement already satisfied: jinja2>=2.10.3 in /usr/local/lib/python3.10/dist-packages (from distributed->ome-zarr>=0.8.4->spatialdata>=0.2.6->spatialdata-plot->crunch-cli) (3.1.4)\n",
            "Requirement already satisfied: msgpack>=1.0.2 in /usr/local/lib/python3.10/dist-packages (from distributed->ome-zarr>=0.8.4->spatialdata>=0.2.6->spatialdata-plot->crunch-cli) (1.1.0)\n",
            "Collecting sortedcontainers>=2.0.5 (from distributed->ome-zarr>=0.8.4->spatialdata>=0.2.6->spatialdata-plot->crunch-cli)\n",
            "  Downloading sortedcontainers-2.4.0-py2.py3-none-any.whl.metadata (10 kB)\n",
            "Collecting tblib>=1.6.0 (from distributed->ome-zarr>=0.8.4->spatialdata>=0.2.6->spatialdata-plot->crunch-cli)\n",
            "  Downloading tblib-3.0.0-py3-none-any.whl.metadata (25 kB)\n",
            "Requirement already satisfied: tornado>=6.2.0 in /usr/local/lib/python3.10/dist-packages (from distributed->ome-zarr>=0.8.4->spatialdata>=0.2.6->spatialdata-plot->crunch-cli) (6.3.3)\n",
            "Collecting zict>=3.0.0 (from distributed->ome-zarr>=0.8.4->spatialdata>=0.2.6->spatialdata-plot->crunch-cli)\n",
            "  Downloading zict-3.0.0-py2.py3-none-any.whl.metadata (899 bytes)\n",
            "INFO: pip is looking at multiple versions of dask-expr to determine which version is compatible with other requirements. This could take a while.\n",
            "Collecting dask-expr<1.2,>=1.1 (from dask[array,dataframe]>=2024.4.1->dask-image->spatialdata>=0.2.6->spatialdata-plot->crunch-cli)\n",
            "  Downloading dask_expr-1.1.20-py3-none-any.whl.metadata (2.6 kB)\n",
            "  Downloading dask_expr-1.1.19-py3-none-any.whl.metadata (2.6 kB)\n",
            "  Downloading dask_expr-1.1.18-py3-none-any.whl.metadata (2.6 kB)\n",
            "  Downloading dask_expr-1.1.16-py3-none-any.whl.metadata (2.5 kB)\n",
            "Requirement already satisfied: MarkupSafe>=2.0 in /usr/local/lib/python3.10/dist-packages (from jinja2>=2.10.3->distributed->ome-zarr>=0.8.4->spatialdata>=0.2.6->spatialdata-plot->crunch-cli) (3.0.2)\n",
            "Collecting aiobotocore<3.0.0,>=2.5.4 (from s3fs->fsspec[s3]!=2021.07.0,>=0.8->ome-zarr>=0.8.4->spatialdata>=0.2.6->spatialdata-plot->crunch-cli)\n",
            "  Downloading aiobotocore-2.16.1-py3-none-any.whl.metadata (23 kB)\n",
            "INFO: pip is looking at multiple versions of s3fs to determine which version is compatible with other requirements. This could take a while.\n",
            "Collecting s3fs (from fsspec[s3]!=2021.07.0,>=0.8->ome-zarr>=0.8.4->spatialdata>=0.2.6->spatialdata-plot->crunch-cli)\n",
            "  Downloading s3fs-2024.10.0-py3-none-any.whl.metadata (1.7 kB)\n",
            "  Downloading s3fs-2024.9.0-py3-none-any.whl.metadata (1.6 kB)\n",
            "  Downloading s3fs-2024.6.1-py3-none-any.whl.metadata (1.6 kB)\n",
            "  Downloading s3fs-2024.6.0-py3-none-any.whl.metadata (1.6 kB)\n",
            "  Downloading s3fs-2024.5.0-py3-none-any.whl.metadata (1.6 kB)\n",
            "  Downloading s3fs-2024.3.1-py3-none-any.whl.metadata (1.6 kB)\n",
            "  Downloading s3fs-2024.3.0-py3-none-any.whl.metadata (1.6 kB)\n",
            "INFO: pip is still looking at multiple versions of s3fs to determine which version is compatible with other requirements. This could take a while.\n",
            "  Downloading s3fs-2024.2.0-py3-none-any.whl.metadata (1.6 kB)\n",
            "  Downloading s3fs-2023.12.2-py3-none-any.whl.metadata (1.6 kB)\n",
            "  Downloading s3fs-2023.12.1-py3-none-any.whl.metadata (1.6 kB)\n",
            "  Downloading s3fs-2023.10.0-py3-none-any.whl.metadata (1.6 kB)\n",
            "Collecting aiobotocore~=2.7.0 (from s3fs->fsspec[s3]!=2021.07.0,>=0.8->ome-zarr>=0.8.4->spatialdata>=0.2.6->spatialdata-plot->crunch-cli)\n",
            "  Downloading aiobotocore-2.7.0-py3-none-any.whl.metadata (20 kB)\n",
            "Collecting s3fs (from fsspec[s3]!=2021.07.0,>=0.8->ome-zarr>=0.8.4->spatialdata>=0.2.6->spatialdata-plot->crunch-cli)\n",
            "  Downloading s3fs-2023.9.2-py3-none-any.whl.metadata (1.6 kB)\n",
            "Collecting aiobotocore~=2.5.4 (from s3fs->fsspec[s3]!=2021.07.0,>=0.8->ome-zarr>=0.8.4->spatialdata>=0.2.6->spatialdata-plot->crunch-cli)\n",
            "  Downloading aiobotocore-2.5.4-py3-none-any.whl.metadata (19 kB)\n",
            "INFO: This is taking longer than usual. You might need to provide the dependency resolver with stricter constraints to reduce runtime. See https://pip.pypa.io/warnings/backtracking for guidance. If you want to abort this run, press Ctrl + C.\n",
            "Collecting s3fs (from fsspec[s3]!=2021.07.0,>=0.8->ome-zarr>=0.8.4->spatialdata>=0.2.6->spatialdata-plot->crunch-cli)\n",
            "  Downloading s3fs-2023.9.1-py3-none-any.whl.metadata (1.6 kB)\n",
            "  Downloading s3fs-2023.9.0-py3-none-any.whl.metadata (1.6 kB)\n",
            "  Downloading s3fs-2023.6.0-py3-none-any.whl.metadata (1.6 kB)\n",
            "Collecting botocore<1.31.18,>=1.31.17 (from aiobotocore~=2.5.4->s3fs->fsspec[s3]!=2021.07.0,>=0.8->ome-zarr>=0.8.4->spatialdata>=0.2.6->spatialdata-plot->crunch-cli)\n",
            "  Downloading botocore-1.31.17-py3-none-any.whl.metadata (5.9 kB)\n",
            "Requirement already satisfied: wrapt<2.0.0,>=1.10.10 in /usr/local/lib/python3.10/dist-packages (from aiobotocore~=2.5.4->s3fs->fsspec[s3]!=2021.07.0,>=0.8->ome-zarr>=0.8.4->spatialdata>=0.2.6->spatialdata-plot->crunch-cli) (1.17.0)\n",
            "Collecting aioitertools<1.0.0,>=0.5.1 (from aiobotocore~=2.5.4->s3fs->fsspec[s3]!=2021.07.0,>=0.8->ome-zarr>=0.8.4->spatialdata>=0.2.6->spatialdata-plot->crunch-cli)\n",
            "  Downloading aioitertools-0.12.0-py3-none-any.whl.metadata (3.8 kB)\n",
            "Collecting jmespath<2.0.0,>=0.7.1 (from botocore<1.31.18,>=1.31.17->aiobotocore~=2.5.4->s3fs->fsspec[s3]!=2021.07.0,>=0.8->ome-zarr>=0.8.4->spatialdata>=0.2.6->spatialdata-plot->crunch-cli)\n",
            "  Downloading jmespath-1.0.1-py3-none-any.whl.metadata (7.6 kB)\n",
            "Collecting urllib3<3,>=1.21.1 (from requests->crunch-cli)\n",
            "  Downloading urllib3-1.26.20-py2.py3-none-any.whl.metadata (50 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m50.1/50.1 kB\u001b[0m \u001b[31m3.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading crunch_cli-5.7.5-py3-none-any.whl (104 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m104.6/104.6 kB\u001b[0m \u001b[31m7.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading coloredlogs-15.0.1-py2.py3-none-any.whl (46 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m46.0/46.0 kB\u001b[0m \u001b[31m2.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading dataclasses_json-0.6.7-py3-none-any.whl (28 kB)\n",
            "Downloading inquirer-3.4.0-py3-none-any.whl (18 kB)\n",
            "Downloading python_dotenv-1.0.1-py3-none-any.whl (19 kB)\n",
            "Downloading redbaron-0.9.2-py2.py3-none-any.whl (34 kB)\n",
            "Downloading spatialdata_plot-0.2.8-py3-none-any.whl (43 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m43.9/43.9 kB\u001b[0m \u001b[31m3.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading baron-0.10.1-py2.py3-none-any.whl (45 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m45.6/45.6 kB\u001b[0m \u001b[31m3.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading blessed-1.20.0-py2.py3-none-any.whl (58 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m58.4/58.4 kB\u001b[0m \u001b[31m4.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading editor-1.6.6-py3-none-any.whl (4.0 kB)\n",
            "Downloading humanfriendly-10.0-py2.py3-none-any.whl (86 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m86.8/86.8 kB\u001b[0m \u001b[31m6.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading marshmallow-3.23.2-py3-none-any.whl (49 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m49.3/49.3 kB\u001b[0m \u001b[31m1.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading readchar-4.2.1-py3-none-any.whl (9.3 kB)\n",
            "Downloading spatialdata-0.2.6-py3-none-any.whl (168 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m168.7/168.7 kB\u001b[0m \u001b[31m12.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading typing_inspect-0.9.0-py3-none-any.whl (8.8 kB)\n",
            "Downloading matplotlib_scalebar-0.8.1-py2.py3-none-any.whl (17 kB)\n",
            "Downloading scanpy-1.10.4-py3-none-any.whl (2.1 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m2.1/2.1 MB\u001b[0m \u001b[31m16.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading anndata-0.11.1-py3-none-any.whl (141 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m141.9/141.9 kB\u001b[0m \u001b[31m8.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading fsspec-2023.6.0-py3-none-any.whl (163 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m163.8/163.8 kB\u001b[0m \u001b[31m8.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading legacy_api_wrap-1.4.1-py3-none-any.whl (10.0 kB)\n",
            "Downloading multiscale_spatial_image-2.0.2-py3-none-any.whl (29 kB)\n",
            "Downloading mypy_extensions-1.0.0-py3-none-any.whl (4.7 kB)\n",
            "Downloading ome_zarr-0.10.2-py3-none-any.whl (37 kB)\n",
            "Downloading pynndescent-0.5.13-py3-none-any.whl (56 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m56.9/56.9 kB\u001b[0m \u001b[31m4.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading spatial_image-1.1.0-py3-none-any.whl (8.0 kB)\n",
            "Downloading umap_learn-0.5.7-py3-none-any.whl (88 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m88.8/88.8 kB\u001b[0m \u001b[31m7.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading xarray_spatial-0.4.0-py3-none-any.whl (2.0 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m2.0/2.0 MB\u001b[0m \u001b[31m51.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading zarr-2.18.3-py3-none-any.whl (210 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m210.7/210.7 kB\u001b[0m \u001b[31m15.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading dask_image-2024.5.3-py3-none-any.whl (59 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m59.5/59.5 kB\u001b[0m \u001b[31m4.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading rply-0.7.8-py2.py3-none-any.whl (16 kB)\n",
            "Downloading runs-1.2.2-py3-none-any.whl (7.0 kB)\n",
            "Downloading xarray_schema-0.0.3-py3-none-any.whl (10 kB)\n",
            "Downloading xmod-1.8.1-py3-none-any.whl (4.6 kB)\n",
            "Downloading array_api_compat-1.10.0-py3-none-any.whl (50 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m50.4/50.4 kB\u001b[0m \u001b[31m3.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading datashader-0.16.3-py2.py3-none-any.whl (18.3 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m18.3/18.3 MB\u001b[0m \u001b[31m36.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading numcodecs-0.13.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (8.6 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m8.6/8.6 MB\u001b[0m \u001b[31m53.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading xarray_dataclasses-1.9.1-py3-none-any.whl (14 kB)\n",
            "Downloading appdirs-1.4.4-py2.py3-none-any.whl (9.6 kB)\n",
            "Downloading distributed-2024.10.0-py3-none-any.whl (1.0 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.0/1.0 MB\u001b[0m \u001b[31m23.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading fasteners-0.19-py3-none-any.whl (18 kB)\n",
            "Downloading stdlib_list-0.11.0-py3-none-any.whl (83 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m83.6/83.6 kB\u001b[0m \u001b[31m6.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading dask_expr-1.1.16-py3-none-any.whl (243 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m243.2/243.2 kB\u001b[0m \u001b[31m17.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading slicerator-1.1.0-py3-none-any.whl (10 kB)\n",
            "Downloading sortedcontainers-2.4.0-py2.py3-none-any.whl (29 kB)\n",
            "Downloading tblib-3.0.0-py3-none-any.whl (12 kB)\n",
            "Downloading zict-3.0.0-py2.py3-none-any.whl (43 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m43.3/43.3 kB\u001b[0m \u001b[31m3.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading pyct-0.5.0-py2.py3-none-any.whl (15 kB)\n",
            "Downloading s3fs-2023.6.0-py3-none-any.whl (28 kB)\n",
            "Downloading aiobotocore-2.5.4-py3-none-any.whl (73 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m73.4/73.4 kB\u001b[0m \u001b[31m3.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading aioitertools-0.12.0-py3-none-any.whl (24 kB)\n",
            "Downloading botocore-1.31.17-py3-none-any.whl (11.1 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m11.1/11.1 MB\u001b[0m \u001b[31m28.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading urllib3-1.26.20-py2.py3-none-any.whl (144 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m144.2/144.2 kB\u001b[0m \u001b[31m10.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading jmespath-1.0.1-py3-none-any.whl (20 kB)\n",
            "Building wheels for collected packages: session-info, pims, asciitree\n",
            "  Building wheel for session-info (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for session-info: filename=session_info-1.0.0-py3-none-any.whl size=8023 sha256=b85cc7c23e790cca8092ba2bed338468688d1d94455dcd24209be4c7a3354b0f\n",
            "  Stored in directory: /root/.cache/pip/wheels/6a/aa/b9/eb5d4031476ec10802795b97ccf937b9bd998d68a9b268765a\n",
            "  Building wheel for pims (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for pims: filename=PIMS-0.7-py3-none-any.whl size=84591 sha256=c61b3817555573eb41015a4012b079e0d199b5a62fce818a5566a287e86e47fb\n",
            "  Stored in directory: /root/.cache/pip/wheels/54/cb/f1/939f4adc0c5bcb1a1a78566d67869368d3d8dc8abd84f63c38\n",
            "  Building wheel for asciitree (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for asciitree: filename=asciitree-0.3.3-py3-none-any.whl size=5034 sha256=75e53f41c30182e9ef3304b544ee332e41439578dc258a1821787d6ee8d6f16b\n",
            "  Stored in directory: /root/.cache/pip/wheels/7f/4e/be/1171b40f43b918087657ec57cf3b81fa1a2e027d8755baa184\n",
            "Successfully built session-info pims asciitree\n",
            "Installing collected packages: sortedcontainers, slicerator, asciitree, appdirs, zict, xmod, urllib3, tblib, stdlib_list, rply, readchar, python-dotenv, pyct, numcodecs, mypy-extensions, marshmallow, legacy-api-wrap, jmespath, humanfriendly, fsspec, fasteners, blessed, array-api-compat, aioitertools, zarr, typing-inspect, session-info, runs, pims, coloredlogs, botocore, baron, redbaron, pynndescent, matplotlib-scalebar, editor, distributed, dataclasses_json, dask-expr, anndata, xarray-schema, xarray-dataclasses, umap-learn, inquirer, datashader, aiobotocore, xarray-spatial, spatial-image, scanpy, s3fs, dask-image, multiscale-spatial-image, ome-zarr, spatialdata, spatialdata-plot, crunch-cli\n",
            "  Attempting uninstall: urllib3\n",
            "    Found existing installation: urllib3 2.2.3\n",
            "    Uninstalling urllib3-2.2.3:\n",
            "      Successfully uninstalled urllib3-2.2.3\n",
            "  Attempting uninstall: fsspec\n",
            "    Found existing installation: fsspec 2024.10.0\n",
            "    Uninstalling fsspec-2024.10.0:\n",
            "      Successfully uninstalled fsspec-2024.10.0\n",
            "\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
            "gcsfs 2024.10.0 requires fsspec==2024.10.0, but you have fsspec 2023.6.0 which is incompatible.\u001b[0m\u001b[31m\n",
            "\u001b[0mSuccessfully installed aiobotocore-2.5.4 aioitertools-0.12.0 anndata-0.11.1 appdirs-1.4.4 array-api-compat-1.10.0 asciitree-0.3.3 baron-0.10.1 blessed-1.20.0 botocore-1.31.17 coloredlogs-15.0.1 crunch-cli-5.7.5 dask-expr-1.1.16 dask-image-2024.5.3 dataclasses_json-0.6.7 datashader-0.16.3 distributed-2024.10.0 editor-1.6.6 fasteners-0.19 fsspec-2023.6.0 humanfriendly-10.0 inquirer-3.4.0 jmespath-1.0.1 legacy-api-wrap-1.4.1 marshmallow-3.23.2 matplotlib-scalebar-0.8.1 multiscale-spatial-image-2.0.2 mypy-extensions-1.0.0 numcodecs-0.13.1 ome-zarr-0.10.2 pims-0.7 pyct-0.5.0 pynndescent-0.5.13 python-dotenv-1.0.1 readchar-4.2.1 redbaron-0.9.2 rply-0.7.8 runs-1.2.2 s3fs-2023.6.0 scanpy-1.10.4 session-info-1.0.0 slicerator-1.1.0 sortedcontainers-2.4.0 spatial-image-1.1.0 spatialdata-0.2.6 spatialdata-plot-0.2.8 stdlib_list-0.11.0 tblib-3.0.0 typing-inspect-0.9.0 umap-learn-0.5.7 urllib3-1.26.20 xarray-dataclasses-1.9.1 xarray-schema-0.0.3 xarray-spatial-0.4.0 xmod-1.8.1 zarr-2.18.3 zict-3.0.0\n",
            "\n",
            "---\n",
            "Your token seems to have expired or is invalid.\n",
            "\n",
            "Please follow this link to copy and paste your new setup command:\n",
            "https://hub.crunchdao.com/competitions/broad-1/submit\n",
            "\n",
            "If you think that is an error, please contact an administrator.\n"
          ]
        }
      ],
      "source": [
        "%pip install --upgrade crunch-cli\n",
        "!crunch setup --notebook broad-1 hello --token aaaabbbbccccddddeeeeffff"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "e4gfY7jC5f-n"
      },
      "source": [
        "## Imports"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "tAzquCA99X33",
        "outputId": "c32f4b3f-9857-4302-e74f-4059a1d82ccd"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "loaded inline runner with module: <module '__main__'>\n"
          ]
        }
      ],
      "source": [
        "import crunch\n",
        "crunch = crunch.load_notebook()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "collapsed": true,
        "id": "0b6AgVirxijj",
        "outputId": "0a98ba09-355e-4558-f55f-c2cd8120912a"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Requirement already satisfied: numpy in /usr/local/lib/python3.10/dist-packages (1.26.4)\n",
            "Requirement already satisfied: pandas in /usr/local/lib/python3.10/dist-packages (2.2.2)\n",
            "Requirement already satisfied: torch in /usr/local/lib/python3.10/dist-packages (2.5.1+cu121)\n",
            "Requirement already satisfied: scipy in /usr/local/lib/python3.10/dist-packages (1.13.1)\n",
            "Collecting openslide-python\n",
            "  Downloading openslide_python-1.4.1-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.whl.metadata (4.3 kB)\n",
            "Requirement already satisfied: pydantic in /usr/local/lib/python3.10/dist-packages (2.10.3)\n",
            "Collecting pytorch-lightning\n",
            "  Downloading pytorch_lightning-2.5.0.post0-py3-none-any.whl.metadata (21 kB)\n",
            "Requirement already satisfied: dask in /usr/local/lib/python3.10/dist-packages (2024.10.0)\n",
            "Requirement already satisfied: distributed in /usr/local/lib/python3.10/dist-packages (2024.10.0)\n",
            "Requirement already satisfied: matplotlib in /usr/local/lib/python3.10/dist-packages (3.8.0)\n",
            "Requirement already satisfied: seaborn in /usr/local/lib/python3.10/dist-packages (0.13.2)\n",
            "Requirement already satisfied: scikit-learn in /usr/local/lib/python3.10/dist-packages (1.6.0)\n",
            "Requirement already satisfied: opencv-python in /usr/local/lib/python3.10/dist-packages (4.10.0.84)\n",
            "Requirement already satisfied: scanpy in /usr/local/lib/python3.10/dist-packages (1.10.4)\n",
            "Requirement already satisfied: spatialdata in /usr/local/lib/python3.10/dist-packages (0.2.6)\n",
            "Requirement already satisfied: zarr in /usr/local/lib/python3.10/dist-packages (2.18.3)\n",
            "Requirement already satisfied: ome-zarr in /usr/local/lib/python3.10/dist-packages (0.10.2)\n",
            "Requirement already satisfied: python-dateutil>=2.8.2 in /usr/local/lib/python3.10/dist-packages (from pandas) (2.8.2)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.10/dist-packages (from pandas) (2024.2)\n",
            "Requirement already satisfied: tzdata>=2022.7 in /usr/local/lib/python3.10/dist-packages (from pandas) (2024.2)\n",
            "Requirement already satisfied: filelock in /usr/local/lib/python3.10/dist-packages (from torch) (3.16.1)\n",
            "Requirement already satisfied: typing-extensions>=4.8.0 in /usr/local/lib/python3.10/dist-packages (from torch) (4.12.2)\n",
            "Requirement already satisfied: networkx in /usr/local/lib/python3.10/dist-packages (from torch) (3.4.2)\n",
            "Requirement already satisfied: jinja2 in /usr/local/lib/python3.10/dist-packages (from torch) (3.1.4)\n",
            "Requirement already satisfied: fsspec in /usr/local/lib/python3.10/dist-packages (from torch) (2023.6.0)\n",
            "Requirement already satisfied: sympy==1.13.1 in /usr/local/lib/python3.10/dist-packages (from torch) (1.13.1)\n",
            "Requirement already satisfied: mpmath<1.4,>=1.1.0 in /usr/local/lib/python3.10/dist-packages (from sympy==1.13.1->torch) (1.3.0)\n",
            "Requirement already satisfied: Pillow in /usr/local/lib/python3.10/dist-packages (from openslide-python) (11.0.0)\n",
            "Requirement already satisfied: annotated-types>=0.6.0 in /usr/local/lib/python3.10/dist-packages (from pydantic) (0.7.0)\n",
            "Requirement already satisfied: pydantic-core==2.27.1 in /usr/local/lib/python3.10/dist-packages (from pydantic) (2.27.1)\n",
            "Requirement already satisfied: tqdm>=4.57.0 in /usr/local/lib/python3.10/dist-packages (from pytorch-lightning) (4.67.1)\n",
            "Requirement already satisfied: PyYAML>=5.4 in /usr/local/lib/python3.10/dist-packages (from pytorch-lightning) (6.0.2)\n",
            "Collecting torchmetrics>=0.7.0 (from pytorch-lightning)\n",
            "  Downloading torchmetrics-1.6.1-py3-none-any.whl.metadata (21 kB)\n",
            "Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.10/dist-packages (from pytorch-lightning) (24.2)\n",
            "Collecting lightning-utilities>=0.10.0 (from pytorch-lightning)\n",
            "  Downloading lightning_utilities-0.11.9-py3-none-any.whl.metadata (5.2 kB)\n",
            "Requirement already satisfied: click>=8.1 in /usr/local/lib/python3.10/dist-packages (from dask) (8.1.7)\n",
            "Requirement already satisfied: cloudpickle>=3.0.0 in /usr/local/lib/python3.10/dist-packages (from dask) (3.1.0)\n",
            "Requirement already satisfied: partd>=1.4.0 in /usr/local/lib/python3.10/dist-packages (from dask) (1.4.2)\n",
            "Requirement already satisfied: toolz>=0.10.0 in /usr/local/lib/python3.10/dist-packages (from dask) (0.12.1)\n",
            "Requirement already satisfied: importlib-metadata>=4.13.0 in /usr/local/lib/python3.10/dist-packages (from dask) (8.5.0)\n",
            "Requirement already satisfied: locket>=1.0.0 in /usr/local/lib/python3.10/dist-packages (from distributed) (1.0.0)\n",
            "Requirement already satisfied: msgpack>=1.0.2 in /usr/local/lib/python3.10/dist-packages (from distributed) (1.1.0)\n",
            "Requirement already satisfied: psutil>=5.8.0 in /usr/local/lib/python3.10/dist-packages (from distributed) (5.9.5)\n",
            "Requirement already satisfied: sortedcontainers>=2.0.5 in /usr/local/lib/python3.10/dist-packages (from distributed) (2.4.0)\n",
            "Requirement already satisfied: tblib>=1.6.0 in /usr/local/lib/python3.10/dist-packages (from distributed) (3.0.0)\n",
            "Requirement already satisfied: tornado>=6.2.0 in /usr/local/lib/python3.10/dist-packages (from distributed) (6.3.3)\n",
            "Requirement already satisfied: urllib3>=1.26.5 in /usr/local/lib/python3.10/dist-packages (from distributed) (1.26.20)\n",
            "Requirement already satisfied: zict>=3.0.0 in /usr/local/lib/python3.10/dist-packages (from distributed) (3.0.0)\n",
            "Requirement already satisfied: contourpy>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib) (1.3.1)\n",
            "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.10/dist-packages (from matplotlib) (0.12.1)\n",
            "Requirement already satisfied: fonttools>=4.22.0 in /usr/local/lib/python3.10/dist-packages (from matplotlib) (4.55.3)\n",
            "Requirement already satisfied: kiwisolver>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib) (1.4.7)\n",
            "Requirement already satisfied: pyparsing>=2.3.1 in /usr/local/lib/python3.10/dist-packages (from matplotlib) (3.2.0)\n",
            "Requirement already satisfied: joblib>=1.2.0 in /usr/local/lib/python3.10/dist-packages (from scikit-learn) (1.4.2)\n",
            "Requirement already satisfied: threadpoolctl>=3.1.0 in /usr/local/lib/python3.10/dist-packages (from scikit-learn) (3.5.0)\n",
            "Requirement already satisfied: anndata>=0.8 in /usr/local/lib/python3.10/dist-packages (from scanpy) (0.11.1)\n",
            "Requirement already satisfied: h5py>=3.6 in /usr/local/lib/python3.10/dist-packages (from scanpy) (3.12.1)\n",
            "Requirement already satisfied: legacy-api-wrap>=1.4 in /usr/local/lib/python3.10/dist-packages (from scanpy) (1.4.1)\n",
            "Requirement already satisfied: natsort in /usr/local/lib/python3.10/dist-packages (from scanpy) (8.4.0)\n",
            "Requirement already satisfied: numba>=0.56 in /usr/local/lib/python3.10/dist-packages (from scanpy) (0.60.0)\n",
            "Requirement already satisfied: patsy!=1.0.0 in /usr/local/lib/python3.10/dist-packages (from scanpy) (1.0.1)\n",
            "Requirement already satisfied: pynndescent>=0.5 in /usr/local/lib/python3.10/dist-packages (from scanpy) (0.5.13)\n",
            "Requirement already satisfied: session-info in /usr/local/lib/python3.10/dist-packages (from scanpy) (1.0.0)\n",
            "Requirement already satisfied: statsmodels>=0.13 in /usr/local/lib/python3.10/dist-packages (from scanpy) (0.14.4)\n",
            "Requirement already satisfied: umap-learn!=0.5.0,>=0.5 in /usr/local/lib/python3.10/dist-packages (from scanpy) (0.5.7)\n",
            "Requirement already satisfied: dask-image in /usr/local/lib/python3.10/dist-packages (from spatialdata) (2024.5.3)\n",
            "Requirement already satisfied: geopandas>=0.14 in /usr/local/lib/python3.10/dist-packages (from spatialdata) (1.0.1)\n",
            "Requirement already satisfied: multiscale-spatial-image>=2.0.2 in /usr/local/lib/python3.10/dist-packages (from spatialdata) (2.0.2)\n",
            "Requirement already satisfied: pooch in /usr/local/lib/python3.10/dist-packages (from spatialdata) (1.8.2)\n",
            "Requirement already satisfied: pyarrow in /usr/local/lib/python3.10/dist-packages (from spatialdata) (17.0.0)\n",
            "Requirement already satisfied: rich in /usr/local/lib/python3.10/dist-packages (from spatialdata) (13.9.4)\n",
            "Requirement already satisfied: scikit-image in /usr/local/lib/python3.10/dist-packages (from spatialdata) (0.25.0)\n",
            "Requirement already satisfied: setuptools in /usr/local/lib/python3.10/dist-packages (from spatialdata) (75.1.0)\n",
            "Requirement already satisfied: shapely>=2.0.1 in /usr/local/lib/python3.10/dist-packages (from spatialdata) (2.0.6)\n",
            "Requirement already satisfied: spatial-image>=1.1.0 in /usr/local/lib/python3.10/dist-packages (from spatialdata) (1.1.0)\n",
            "Requirement already satisfied: xarray-schema in /usr/local/lib/python3.10/dist-packages (from spatialdata) (0.0.3)\n",
            "Requirement already satisfied: xarray-spatial>=0.3.5 in /usr/local/lib/python3.10/dist-packages (from spatialdata) (0.4.0)\n",
            "Requirement already satisfied: xarray>=2024.10.0 in /usr/local/lib/python3.10/dist-packages (from spatialdata) (2024.11.0)\n",
            "Requirement already satisfied: asciitree in /usr/local/lib/python3.10/dist-packages (from zarr) (0.3.3)\n",
            "Requirement already satisfied: numcodecs>=0.10.0 in /usr/local/lib/python3.10/dist-packages (from zarr) (0.13.1)\n",
            "Requirement already satisfied: fasteners in /usr/local/lib/python3.10/dist-packages (from zarr) (0.19)\n",
            "Requirement already satisfied: aiohttp<4 in /usr/local/lib/python3.10/dist-packages (from ome-zarr) (3.11.10)\n",
            "Requirement already satisfied: requests in /usr/local/lib/python3.10/dist-packages (from ome-zarr) (2.32.3)\n",
            "Requirement already satisfied: aiohappyeyeballs>=2.3.0 in /usr/local/lib/python3.10/dist-packages (from aiohttp<4->ome-zarr) (2.4.4)\n",
            "Requirement already satisfied: aiosignal>=1.1.2 in /usr/local/lib/python3.10/dist-packages (from aiohttp<4->ome-zarr) (1.3.2)\n",
            "Requirement already satisfied: async-timeout<6.0,>=4.0 in /usr/local/lib/python3.10/dist-packages (from aiohttp<4->ome-zarr) (4.0.3)\n",
            "Requirement already satisfied: attrs>=17.3.0 in /usr/local/lib/python3.10/dist-packages (from aiohttp<4->ome-zarr) (24.3.0)\n",
            "Requirement already satisfied: frozenlist>=1.1.1 in /usr/local/lib/python3.10/dist-packages (from aiohttp<4->ome-zarr) (1.5.0)\n",
            "Requirement already satisfied: multidict<7.0,>=4.5 in /usr/local/lib/python3.10/dist-packages (from aiohttp<4->ome-zarr) (6.1.0)\n",
            "Requirement already satisfied: propcache>=0.2.0 in /usr/local/lib/python3.10/dist-packages (from aiohttp<4->ome-zarr) (0.2.1)\n",
            "Requirement already satisfied: yarl<2.0,>=1.17.0 in /usr/local/lib/python3.10/dist-packages (from aiohttp<4->ome-zarr) (1.18.3)\n",
            "Requirement already satisfied: array-api-compat!=1.5,>1.4 in /usr/local/lib/python3.10/dist-packages (from anndata>=0.8->scanpy) (1.10.0)\n",
            "Requirement already satisfied: exceptiongroup in /usr/local/lib/python3.10/dist-packages (from anndata>=0.8->scanpy) (1.2.2)\n",
            "Requirement already satisfied: s3fs in /usr/local/lib/python3.10/dist-packages (from fsspec[s3]!=2021.07.0,>=0.8->ome-zarr) (2023.6.0)\n",
            "Requirement already satisfied: pyogrio>=0.7.2 in /usr/local/lib/python3.10/dist-packages (from geopandas>=0.14->spatialdata) (0.10.0)\n",
            "Requirement already satisfied: pyproj>=3.3.0 in /usr/local/lib/python3.10/dist-packages (from geopandas>=0.14->spatialdata) (3.7.0)\n",
            "Requirement already satisfied: zipp>=3.20 in /usr/local/lib/python3.10/dist-packages (from importlib-metadata>=4.13.0->dask) (3.21.0)\n",
            "Requirement already satisfied: MarkupSafe>=2.0 in /usr/local/lib/python3.10/dist-packages (from jinja2->torch) (3.0.2)\n",
            "Requirement already satisfied: llvmlite<0.44,>=0.43.0dev0 in /usr/local/lib/python3.10/dist-packages (from numba>=0.56->scanpy) (0.43.0)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.10/dist-packages (from python-dateutil>=2.8.2->pandas) (1.17.0)\n",
            "Requirement already satisfied: xarray-dataclasses>=1.1.0 in /usr/local/lib/python3.10/dist-packages (from spatial-image>=1.1.0->spatialdata) (1.9.1)\n",
            "Requirement already satisfied: datashader>=0.15.0 in /usr/local/lib/python3.10/dist-packages (from xarray-spatial>=0.3.5->spatialdata) (0.16.3)\n",
            "Requirement already satisfied: pims>=0.4.1 in /usr/local/lib/python3.10/dist-packages (from dask-image->spatialdata) (0.7)\n",
            "Requirement already satisfied: tifffile>=2018.10.18 in /usr/local/lib/python3.10/dist-packages (from dask-image->spatialdata) (2024.12.12)\n",
            "Requirement already satisfied: platformdirs>=2.5.0 in /usr/local/lib/python3.10/dist-packages (from pooch->spatialdata) (4.3.6)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.10/dist-packages (from requests->ome-zarr) (3.4.0)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.10/dist-packages (from requests->ome-zarr) (3.10)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.10/dist-packages (from requests->ome-zarr) (2024.12.14)\n",
            "Requirement already satisfied: markdown-it-py>=2.2.0 in /usr/local/lib/python3.10/dist-packages (from rich->spatialdata) (3.0.0)\n",
            "Requirement already satisfied: pygments<3.0.0,>=2.13.0 in /usr/local/lib/python3.10/dist-packages (from rich->spatialdata) (2.18.0)\n",
            "Requirement already satisfied: imageio!=2.35.0,>=2.33 in /usr/local/lib/python3.10/dist-packages (from scikit-image->spatialdata) (2.36.1)\n",
            "Requirement already satisfied: lazy-loader>=0.4 in /usr/local/lib/python3.10/dist-packages (from scikit-image->spatialdata) (0.4)\n",
            "Requirement already satisfied: stdlib-list in /usr/local/lib/python3.10/dist-packages (from session-info->scanpy) (0.11.0)\n",
            "Requirement already satisfied: dask-expr<1.2,>=1.1 in /usr/local/lib/python3.10/dist-packages (from dask[array,dataframe]>=2024.4.1->dask-image->spatialdata) (1.1.16)\n",
            "Requirement already satisfied: colorcet in /usr/local/lib/python3.10/dist-packages (from datashader>=0.15.0->xarray-spatial>=0.3.5->spatialdata) (3.1.0)\n",
            "Requirement already satisfied: multipledispatch in /usr/local/lib/python3.10/dist-packages (from datashader>=0.15.0->xarray-spatial>=0.3.5->spatialdata) (1.0.0)\n",
            "Requirement already satisfied: param in /usr/local/lib/python3.10/dist-packages (from datashader>=0.15.0->xarray-spatial>=0.3.5->spatialdata) (2.2.0)\n",
            "Requirement already satisfied: pyct in /usr/local/lib/python3.10/dist-packages (from datashader>=0.15.0->xarray-spatial>=0.3.5->spatialdata) (0.5.0)\n",
            "Requirement already satisfied: mdurl~=0.1 in /usr/local/lib/python3.10/dist-packages (from markdown-it-py>=2.2.0->rich->spatialdata) (0.1.2)\n",
            "Requirement already satisfied: slicerator>=0.9.8 in /usr/local/lib/python3.10/dist-packages (from pims>=0.4.1->dask-image->spatialdata) (1.1.0)\n",
            "Requirement already satisfied: aiobotocore~=2.5.0 in /usr/local/lib/python3.10/dist-packages (from s3fs->fsspec[s3]!=2021.07.0,>=0.8->ome-zarr) (2.5.4)\n",
            "Requirement already satisfied: botocore<1.31.18,>=1.31.17 in /usr/local/lib/python3.10/dist-packages (from aiobotocore~=2.5.0->s3fs->fsspec[s3]!=2021.07.0,>=0.8->ome-zarr) (1.31.17)\n",
            "Requirement already satisfied: wrapt<2.0.0,>=1.10.10 in /usr/local/lib/python3.10/dist-packages (from aiobotocore~=2.5.0->s3fs->fsspec[s3]!=2021.07.0,>=0.8->ome-zarr) (1.17.0)\n",
            "Requirement already satisfied: aioitertools<1.0.0,>=0.5.1 in /usr/local/lib/python3.10/dist-packages (from aiobotocore~=2.5.0->s3fs->fsspec[s3]!=2021.07.0,>=0.8->ome-zarr) (0.12.0)\n",
            "Requirement already satisfied: jmespath<2.0.0,>=0.7.1 in /usr/local/lib/python3.10/dist-packages (from botocore<1.31.18,>=1.31.17->aiobotocore~=2.5.0->s3fs->fsspec[s3]!=2021.07.0,>=0.8->ome-zarr) (1.0.1)\n",
            "Downloading openslide_python-1.4.1-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.whl (36 kB)\n",
            "Downloading pytorch_lightning-2.5.0.post0-py3-none-any.whl (819 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m819.3/819.3 kB\u001b[0m \u001b[31m24.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading lightning_utilities-0.11.9-py3-none-any.whl (28 kB)\n",
            "Downloading torchmetrics-1.6.1-py3-none-any.whl (927 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m927.3/927.3 kB\u001b[0m \u001b[31m26.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: openslide-python, lightning-utilities, torchmetrics, pytorch-lightning\n",
            "Successfully installed lightning-utilities-0.11.9 openslide-python-1.4.1 pytorch-lightning-2.5.0.post0 torchmetrics-1.6.1\n"
          ]
        }
      ],
      "source": [
        "%pip install numpy pandas torch scipy openslide-python pydantic pytorch-lightning dask distributed matplotlib seaborn scikit-learn opencv-python scanpy spatialdata zarr ome-zarr"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "collapsed": true,
        "id": "vd1p0CnExnj5",
        "outputId": "6289b662-cf9b-448a-9306-c067682420c9"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.10/dist-packages/cudf/utils/_ptxcompiler.py:64: UserWarning: Error getting driver and runtime versions:\n",
            "\n",
            "stdout:\n",
            "\n",
            "\n",
            "\n",
            "stderr:\n",
            "\n",
            "Traceback (most recent call last):\n",
            "  File \"<string>\", line 4, in <module>\n",
            "  File \"/usr/local/lib/python3.10/dist-packages/numba/cuda/cudadrv/driver.py\", line 295, in __getattr__\n",
            "    raise CudaSupportError(\"Error at driver init: \\n%s:\" %\n",
            "numba.cuda.cudadrv.error.CudaSupportError: Error at driver init: \n",
            "\n",
            "CUDA driver library cannot be found.\n",
            "If you are sure that a CUDA driver is installed,\n",
            "try setting environment variable NUMBA_CUDA_DRIVER\n",
            "with the file path of the CUDA driver shared library.\n",
            ":\n",
            "\n",
            "\n",
            "Not patching Numba\n",
            "  warnings.warn(msg, UserWarning)\n",
            "/usr/local/lib/python3.10/dist-packages/cudf/utils/gpu_utils.py:62: UserWarning: Failed to dlopen libcuda.so.1\n",
            "  warnings.warn(str(e))\n",
            "/usr/local/lib/python3.10/dist-packages/cudf/utils/gpu_utils.py:62: UserWarning: Function \"cuInit\" not found\n",
            "  warnings.warn(str(e))\n"
          ]
        }
      ],
      "source": [
        "# Import libraries\n",
        "# Core Python Libraries\n",
        "import os\n",
        "import json\n",
        "import warnings\n",
        "import gc\n",
        "import joblib\n",
        "from types import SimpleNamespace\n",
        "from operator import itemgetter\n",
        "from abc import abstractmethod\n",
        "from tqdm import tqdm\n",
        "\n",
        "# Core Data manipulation Libraries\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "from scipy.stats import pearsonr, ConstantInputWarning\n",
        "\n",
        "# Visualization Library\n",
        "import matplotlib.pyplot as plt\n",
        "from matplotlib.collections import PatchCollection\n",
        "from matplotlib.patches import Rectangle\n",
        "\n",
        "# Spatial Data Processing\n",
        "import spatialdata as sd  # Manage multi-modal spatial omics datasets\n",
        "import anndata as ad  # Manage annotated data matrices in memory and on disk\n",
        "import scanpy as sc  # For analyzing single-cell data, especially for dimensionality reduction and clustering.\n",
        "from skimage.measure import regionprops  # Get region properties of nucleus/cell image from masked nucleus image\n",
        "import h5py  # For handling HDF5 data files\n",
        "\n",
        "# Frameworks for ML and DL models\n",
        "import torch\n",
        "import timm  # timm: A library to load pretrained SOTA computer vision models (e.g. classification, feature extraction, ...)\n",
        "from sklearn.linear_model import Ridge  # Regression model"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "db_767G4nCso"
      },
      "source": [
        "## Utilities for Saving and Reading HDF5 Files\n",
        "\n",
        "This section contains functions for efficiently saving and loading data from HDF5 files, which is a common format for storing large datasets.\n",
        "\n",
        "*   `save_hdf5` & `read_assets_from_h5`: Functions for saving and reading datasets and attributes to/from HDF5 files.\n",
        "*   `Patcher` class: Extracts image patches from a larger image using given coordinates.\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "pTzYKCIfENgW"
      },
      "source": [
        "![patcher](https://raw.githubusercontent.com/crunchdao/competitions/refs/heads/master/competitions/broad-1/quickstarters/resnet50-plus-ridge/images/patcher.png)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {
        "id": "n4N7eth0Sfss"
      },
      "outputs": [],
      "source": [
        "#####\n",
        "#  Utilities for saving and reading HDF5 files\n",
        "####\n",
        "\n",
        "def save_hdf5(output_fpath, asset_dict, attr_dict=None, mode='a', auto_chunk=True, chunk_size=None):\n",
        "    \"\"\"\n",
        "    Save data and attributes into an HDF5 file, or initialize a new file with the given data.\n",
        "\n",
        "    Parameters:\n",
        "        output_fpath (str): Path to save the HDF5 file.\n",
        "        asset_dict (dict): Dictionary containing keys and their corresponding data (e.g., numpy arrays) to save.\n",
        "        attr_dict (dict, optional): Dictionary of attributes for each key. Format: {key: {attr_key: attr_val, ...}}.\n",
        "        mode (str): File mode ('a' for append, 'w' for write, etc.).\n",
        "        auto_chunk (bool): Whether to enable automatic chunking for HDF5 datasets.\n",
        "        chunk_size (int, optional): If auto_chunk is False, specify the chunk size for the first dimension.\n",
        "\n",
        "    Returns:\n",
        "        str: Path of the saved HDF5 file.\n",
        "    \"\"\"\n",
        "\n",
        "    with h5py.File(output_fpath, mode) as f:\n",
        "        for key, val in asset_dict.items():\n",
        "            data_shape = val.shape\n",
        "            # Ensure data has at least 2 dimensions\n",
        "            if len(data_shape) == 1:\n",
        "                val = np.expand_dims(val, axis=1)\n",
        "                data_shape = val.shape\n",
        "\n",
        "            if key not in f:  # if key does not exist, create a new dataset\n",
        "                data_type = val.dtype\n",
        "\n",
        "                if data_type.kind == 'U':  # Handle Unicode strings\n",
        "                    chunks = (1, 1)\n",
        "                    max_shape = (None, 1)\n",
        "                    data_type = h5py.string_dtype(encoding='utf-8')\n",
        "                else:\n",
        "                    if data_type == np.object_:\n",
        "                        data_type = h5py.string_dtype(encoding='utf-8')\n",
        "                    # Determine chunking strategy\n",
        "                    if auto_chunk:\n",
        "                        chunks = True  # let h5py decide chunk size\n",
        "                    else:\n",
        "                        chunks = (chunk_size,) + data_shape[1:]\n",
        "                    maxshape = (None,) + data_shape[1:]  # Allow unlimited size for the first dimension\n",
        "\n",
        "                try:\n",
        "                    dset = f.create_dataset(key,\n",
        "                                            shape=data_shape,\n",
        "                                            chunks=chunks,\n",
        "                                            maxshape=maxshape,\n",
        "                                            dtype=data_type)\n",
        "                    # Save attributes for the dataset\n",
        "                    if attr_dict is not None:\n",
        "                        if key in attr_dict.keys():\n",
        "                            for attr_key, attr_val in attr_dict[key].items():\n",
        "                                dset.attrs[attr_key] = attr_val\n",
        "                    # Write the data to the dataset\n",
        "                    dset[:] = val\n",
        "                except:\n",
        "                    print(f\"Error encoding {key} of dtype {data_type} into hdf5\")\n",
        "\n",
        "            else:  # Append data to an existing dataset\n",
        "                dset = f[key]\n",
        "                dset.resize(len(dset) + data_shape[0], axis=0)\n",
        "                # assert dset.dtype == val.dtype\n",
        "                dset[-data_shape[0]:] = val\n",
        "\n",
        "    return output_fpath\n",
        "\n",
        "\n",
        "def read_assets_from_h5(h5_path, keys=None, skip_attrs=False, skip_assets=False):\n",
        "    \"\"\"\n",
        "    Read data and attributes from an HDF5 file.\n",
        "\n",
        "    Parameters:\n",
        "        h5_path (str): Path to the HDF5 file.\n",
        "        keys (list, optional): List of keys to read. Reads all keys if None.\n",
        "        skip_attrs (bool): If True, skip reading attributes.\n",
        "        skip_assets (bool): If True, skip reading data assets.\n",
        "\n",
        "    Returns:\n",
        "        tuple: A dictionary of data assets and a dictionary of attributes.\n",
        "    \"\"\"\n",
        "\n",
        "    assets = {}\n",
        "    attrs = {}\n",
        "    with h5py.File(h5_path, 'r') as f:\n",
        "        if keys is None:\n",
        "            keys = list(f.keys())\n",
        "\n",
        "        for key in keys:\n",
        "            if not skip_assets:\n",
        "                assets[key] = f[key][:]\n",
        "            if not skip_attrs and f[key].attrs is not None:\n",
        "                attrs[key] = dict(f[key].attrs)\n",
        "\n",
        "    return assets, attrs\n",
        "\n",
        "\n",
        "class Patcher:\n",
        "    def __init__(self, image, coords, patch_size_target, name=None):\n",
        "        \"\"\"\n",
        "        Initializes the patcher object to extract patches (localized square sub-region of an image) from an image at specified coordinates.\n",
        "\n",
        "        :param image: Input image as a numpy array (H x W x 3), the input image from which patches will be extracted.\n",
        "        :param coords: List or array of cell coordinates (centroïd) [(x1, y1), (x2, y2), ...].\n",
        "        :param patch_size_target: Target size of patches.\n",
        "        :param name: Name of the whole slide image (optional).\n",
        "        \"\"\"\n",
        "\n",
        "        self.image = image\n",
        "        self.height, self.width = image.shape[:2]\n",
        "        self.coords = coords\n",
        "        self.patch_size_target = patch_size_target\n",
        "        self.name = name\n",
        "\n",
        "    def __iter__(self):\n",
        "        \"\"\"\n",
        "        Iterates over coordinates, yielding image patches and their coordinates.\n",
        "        \"\"\"\n",
        "\n",
        "        for x, y in self.coords:\n",
        "            # Extract patch dimension centered at (x, y)\n",
        "            x_start = max(x - self.patch_size_target // 2, 0)\n",
        "            y_start = max(y - self.patch_size_target // 2, 0)\n",
        "            x_end = min(x_start + self.patch_size_target, self.width)\n",
        "            y_end = min(y_start + self.patch_size_target, self.height)\n",
        "\n",
        "            # Ensure the patch size matches the target size, padding with zeros if necessary\n",
        "            patch = np.zeros((self.patch_size_target, self.patch_size_target, 3), dtype=np.uint8)\n",
        "            patch[:y_end - y_start, :x_end - x_start, :] = self.image[y_start:y_end, x_start:x_end, :]\n",
        "\n",
        "            yield patch, x, y\n",
        "\n",
        "    def __len__(self):\n",
        "        \"\"\"\n",
        "        Returns the number of patches based on the number of coordinates.\n",
        "        This is used to determine how many iterations will be done when iterating over the object.\n",
        "        \"\"\"\n",
        "\n",
        "        return len(self.coords)\n",
        "\n",
        "    def save_visualization(self, path, vis_width=300, dpi=150):\n",
        "        \"\"\"\n",
        "        Save a visualization of patches overlayed on the tissue H&E image.\n",
        "        This function creates a plot where each patch's location is marked with a rectangle overlaid on the image.\n",
        "\n",
        "        :param path: File path where the visualization will be saved.\n",
        "        :param vis_width: Target width of the visualization in pixels.\n",
        "        :param dpi: Resolution of the saved visualization.\n",
        "        \"\"\"\n",
        "\n",
        "        # Generate the tissue visualization mask\n",
        "        mask_plot = self.image\n",
        "\n",
        "        # Calculate downscale factor for visualization\n",
        "        downscale_vis = vis_width / self.width\n",
        "\n",
        "        # Create a plot\n",
        "        _, ax = plt.subplots(figsize=(self.height / self.width * vis_width / dpi, vis_width / dpi))\n",
        "        ax.imshow(mask_plot)\n",
        "\n",
        "        # Add patches\n",
        "        patch_rectangles = []\n",
        "        for x, y in self.coords:\n",
        "            x_start, y_start = x - self.patch_size_target // 2, y - self.patch_size_target // 2\n",
        "            patch_rectangles.append(Rectangle((x_start, y_start), self.patch_size_target, self.patch_size_target))\n",
        "\n",
        "        # Add rectangles to the plot\n",
        "        ax.add_collection(PatchCollection(patch_rectangles, facecolor='none', edgecolor='black', linewidth=0.3))\n",
        "\n",
        "        ax.set_axis_off()\n",
        "        plt.tight_layout()\n",
        "        plt.savefig(path, dpi=dpi, bbox_inches='tight')\n",
        "        plt.show()\n",
        "        plt.close()\n",
        "\n",
        "    def view_coord_points(self, vis_width=300, dpi=150):\n",
        "        \"\"\"\n",
        "        Visualizes the coordinates as small points in 2D.\n",
        "        This function generates a scatter plot of the patch coordinates on the H&E image.\n",
        "        \"\"\"\n",
        "\n",
        "        # Calculate downscale factor for visualization\n",
        "        downscale_vis = vis_width / self.width\n",
        "\n",
        "        # Create a plot\n",
        "        _, ax = plt.subplots(figsize=(self.height / self.width * vis_width / dpi, vis_width / dpi))\n",
        "        plt.scatter(self.coords[:, 0], -self.coords[:, 1], s=0.2)\n",
        "        plt.show()\n",
        "        plt.close()\n",
        "\n",
        "    def to_h5(self, path, extra_assets={}):\n",
        "        \"\"\"\n",
        "        Saves the extracted patches and their associated information to an HDF5 file.\n",
        "\n",
        "        Each patch is saved as a dataset along with its coordinates and any additional assets (extra_assets).\n",
        "        The HDF5 file is structured with a dataset for the image patch ('img') and coordinates ('coords').\n",
        "\n",
        "        :param path: File path where the HDF5 file will be saved.\n",
        "        :param extra_assets: Dictionary of additional assets to save (optional). Each value in extra_assets must have the same length as the patches.\n",
        "        \"\"\"\n",
        "\n",
        "        mode_HE = 'w'  # Start with write mode for the first patch\n",
        "        i = 0\n",
        "\n",
        "        # Check that the extra_assets match the number of patches\n",
        "        if extra_assets:\n",
        "            for _, value in extra_assets.items():\n",
        "                if len(value) != len(self):\n",
        "                    raise ValueError(\"Each value in extra_assets must have the same length as the patcher object.\")\n",
        "\n",
        "        # Ensure the file has the correct extension\n",
        "        if not (path.endswith('.h5') or path.endswith('.h5ad')):\n",
        "            path = path + '.h5'\n",
        "\n",
        "        # Loop through each patch and save it to the HDF5 file (loop through __iter__ function)\n",
        "        for tile, x, y in tqdm(self):\n",
        "            assert tile.shape == (self.patch_size_target, self.patch_size_target, 3)\n",
        "\n",
        "            # Prepare the data to be saved for this patch\n",
        "            asset_dict = {\n",
        "                'img': np.expand_dims(tile, axis=0),  # Shape (1, h, w, 3)\n",
        "                'coords': np.expand_dims([x, y], axis=0)  # Shape (1, 2)\n",
        "            }\n",
        "\n",
        "            # Add any extra assets to the asset dictionary\n",
        "            extra_asset_dict = {key: np.expand_dims([value[i]], axis=0) for key, value in extra_assets.items()}\n",
        "            asset_dict = {**asset_dict, **extra_asset_dict}\n",
        "\n",
        "            # Define the attributes for the image patch\n",
        "            attr_dict = {'img': {'patch_size_target': self.patch_size_target}}\n",
        "\n",
        "            if self.name is not None:\n",
        "                attr_dict['img']['name'] = self.name\n",
        "\n",
        "            # Save the patch data to the HDF5 file\n",
        "            save_hdf5(path, asset_dict, attr_dict, mode=mode_HE, auto_chunk=False, chunk_size=1)\n",
        "            mode_HE = 'a'  # Switch to append mode after the first patch\n",
        "            i += 1"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "oneR4KRvpMbS"
      },
      "source": [
        "## Preprocessing Spatial Transcriptomics Data\n",
        "\n",
        "This section contains functions for preprocessing spatial transcriptomics data, including extracting spatial coordinates, generating image patches and preparing datasets for training and testing.\n",
        "\n",
        "* **`extract_spatial_positions`**: Extracts spatial coordinates (centroids) of cells.\n",
        "* **`process_and_visualize_image`**: Extracts square image patches from H&E images and visualizes them.\n",
        "* **`preprocess_spatial_transcriptomics_data_train`**: Prepares training data by generating gene expression (Y) and image patch datasets (X).\n",
        "* **`preprocess_spatial_transcriptomics_data_test`**: Prepares test data by generating image patches (X) for selected cells.\n",
        "* **`create_cross_validation_splits`**: Creates leave-one-out cross-validation splits for model evaluation.\n",
        "\n",
        "![data_X_Y](https://raw.githubusercontent.com/crunchdao/competitions/refs/heads/master/competitions/broad-1/quickstarters/resnet50-plus-ridge/images/data_X_Y.png)\n",
        "\n",
        "`Leave-one-out cross-validation schema:`\n",
        "![cross_validation](https://raw.githubusercontent.com/crunchdao/competitions/refs/heads/master/competitions/broad-1/quickstarters/resnet50-plus-ridge/images/cross_validation.png)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {
        "id": "DRqmAKSqYtJS"
      },
      "outputs": [],
      "source": [
        "##########\n",
        "# Preprocessing spatial transcriptomics data\n",
        "##########\n",
        "\n",
        "def extract_spatial_positions(sdata, cell_id_list):\n",
        "    \"\"\"\n",
        "    Extracts spatial positions (centroids) of regions from the nucleus image where cell IDs match the provided cell list.\n",
        "\n",
        "    Need to use 'HE_nuc_original' to extract spatial coordinate of cells\n",
        "    HE_nuc_original: The nucleus segmentation mask of H&E image, in H&E native coordinate system. The cell_id in this segmentation mask matches with the nuclei by gene matrix stored in anucleus.\n",
        "    HE_nuc_original is like a binary segmentation mask 0 - 1 but replace 1 with cell_ids.\n",
        "    You can directly find the location of a cell, with cell_id, through HE_nuc_original==cell_id\n",
        "\n",
        "    Parameters:\n",
        "    -----------\n",
        "    sdata: SpatialData\n",
        "        A spatial data object containing the nucleus segmentation mask ('HE_nuc_original').\n",
        "    cell_id_list: array-like\n",
        "        A list or array of cell IDs to filter the regions.\n",
        "\n",
        "    Returns:\n",
        "    --------\n",
        "    np.ndarray\n",
        "        A NumPy array of spatial coordinates (x_center, y_center) for matched regions.\n",
        "    \"\"\"\n",
        "\n",
        "    print(\"Extracting spatial positions ...\")\n",
        "    # Get region properties from the nucleus image: for each cell_id get its location on HE image\n",
        "    regions = regionprops(sdata['HE_nuc_original'][0, :, :].to_numpy())\n",
        "\n",
        "    dict_spatial_positions = {}\n",
        "    # Loop through each region and extract centroid if the cell ID matches\n",
        "    for props in tqdm(regions):\n",
        "        cell_id = props.label\n",
        "        centroid = props.centroid\n",
        "        # Extract only coordinates from the provided cell_id list\n",
        "        if cell_id in cell_id_list:\n",
        "            y_center, x_center = int(centroid[0]), int(centroid[1])\n",
        "            dict_spatial_positions[cell_id] = [x_center, y_center]\n",
        "\n",
        "    # To maintain cell IDs order\n",
        "    spatial_positions = []\n",
        "    for cell_id in cell_id_list:\n",
        "        try:\n",
        "            spatial_positions.append(dict_spatial_positions[cell_id])\n",
        "        except KeyError:\n",
        "            print(f\"Warning: Cell ID {cell_id} not found in the segmentation mask.\")\n",
        "            spatial_positions.append([1000, 1000])\n",
        "\n",
        "    return np.array(spatial_positions)\n",
        "\n",
        "\n",
        "def process_and_visualize_image(sdata, patch_save_dir, name_data, coords_center, target_patch_size, barcodes,\n",
        "                                show_extracted_images=False, vis_width=1000):\n",
        "    \"\"\"\n",
        "    Load and process the spatial image data, creates patches, saves them in an HDF5 file,\n",
        "    and visualizes the extracted images and spatial coordinates.\n",
        "\n",
        "    Parameters:\n",
        "    -----------\n",
        "    sdata: SpatialData\n",
        "        A spatial data object containing the image to process ('HE_original') and associated metadata.\n",
        "    patch_save_dir: str\n",
        "        Directory where the resulting HDF5 file and visualizations will be saved.\n",
        "    name_data: str\n",
        "        Name used for saving the dataset.\n",
        "    coords_center: array-like\n",
        "        Coordinates of the regions to be patched (centroids of cell regions).\n",
        "    target_patch_size: int\n",
        "        Size of the patches to extract from the image.\n",
        "    barcodes: array-like\n",
        "        Barcodes associated with patches.\n",
        "    show_extracted_images: bool, optional (default=False)\n",
        "        If True, will show extracted images during the visualization phase.\n",
        "    vis_width: int, optional (default=1000)\n",
        "        Width of the visualization images.\n",
        "    \"\"\"\n",
        "\n",
        "    # Load the image and transpose it to the correct format\n",
        "    print(\"Loading imgs ...\")\n",
        "    intensity_image = np.transpose(sdata['HE_original'].to_numpy(), (1, 2, 0))\n",
        "\n",
        "    print(\"Patching: create image dataset (X) ...\")\n",
        "    # Path for the .h5 image dataset\n",
        "    h5_path = os.path.join(patch_save_dir, name_data + '.h5')\n",
        "\n",
        "    # Create the patcher object to extract patches (localized square sub-region of an image) from an image at specified coordinates.\n",
        "    patcher = Patcher(\n",
        "        image=intensity_image,\n",
        "        coords=coords_center,\n",
        "        patch_size_target=target_patch_size\n",
        "    )\n",
        "\n",
        "    # Build and Save patches to an HDF5 file\n",
        "    patcher.to_h5(h5_path, extra_assets={'barcode': barcodes})\n",
        "\n",
        "    # Visualization\n",
        "    print(\"Visualization\")\n",
        "    if show_extracted_images:\n",
        "        print(\"Extracted Images (high time and memory consumption...)\")\n",
        "        patcher.save_visualization(os.path.join(patch_save_dir, name_data + '_viz.png'), vis_width=vis_width)\n",
        "\n",
        "    print(\"Spatial coordinates\")\n",
        "    patcher.view_coord_points(vis_width=vis_width)\n",
        "\n",
        "    # Display some example images from the created dataset\n",
        "    print(\"Examples from the created .h5 dataset\")\n",
        "    assets, _ = read_assets_from_h5(h5_path)\n",
        "\n",
        "    n_images = 3\n",
        "    fig, axes = plt.subplots(1, n_images, figsize=(15, 5))\n",
        "    for i in range(n_images):\n",
        "        axes[i].imshow(assets[\"img\"][i])\n",
        "    for ax in axes:\n",
        "        ax.axis('off')\n",
        "    plt.show()\n",
        "\n",
        "    # Delete variables that are no longer used\n",
        "    del intensity_image, patcher, assets\n",
        "    gc.collect()\n",
        "\n",
        "\n",
        "def preprocess_spatial_transcriptomics_data_train(list_ST_name_data, data_directory_path, dir_processed_dataset, size_subset=None, target_patch_size=32, vis_width=1000, show_extracted_images=False):\n",
        "    \"\"\"\n",
        "    Train step: Preprocesses spatial transcriptomics data by performing the following steps for each ST:\n",
        "    1. Samples the dataset and extract spatial coordinates of cells.\n",
        "    2. Extract gene expression data (Y) and save it as `.h5ad` files into directory 'adata'.\n",
        "    4. Generates and saves patches of images centered on spatial coordinates to HDF5 files (X) into directory 'patches'.\n",
        "    5. Saves the list of genes to a JSON file into direcotry 'splits'.\n",
        "\n",
        "    Parameters:\n",
        "    -----------\n",
        "    list_ST_name_data: list\n",
        "        List of spatial transcriptomics data names.\n",
        "    data_directory_path: str\n",
        "        Path to the directory containing the input data in `.zarr` format.\n",
        "    dir_processed_dataset: str\n",
        "        Path to the directory where processed datasets and outputs will be saved.\n",
        "    size_subset: int, optional\n",
        "        ST data sample size. If None, no sampling.\n",
        "    target_patch_size: int, optional\n",
        "        Target size of image patches to extract.\n",
        "    vis_width: int, optional\n",
        "        Width of the visualization output for spatial and image patches.\n",
        "    show_extracted_images: bool\n",
        "    \"\"\"\n",
        "\n",
        "    # Creates directories for saving patches (X) ('patches'), processed AnnData objects (Y) ('adata'), and train/test dataset splits ('splits').\n",
        "    patch_save_dir = os.path.join(dir_processed_dataset, \"patches\")\n",
        "    adata_save_dir = os.path.join(dir_processed_dataset, \"adata\")\n",
        "    splits_save_dir = os.path.join(dir_processed_dataset, \"splits\")\n",
        "    os.makedirs(patch_save_dir, exist_ok=True)\n",
        "    os.makedirs(adata_save_dir, exist_ok=True)\n",
        "    os.makedirs(splits_save_dir, exist_ok=True)\n",
        "\n",
        "    print(\"\\n -- PREPROCESS SPATIAL TRANSCRIPTOMICS DATASET --------------------------------------------\\n\")\n",
        "\n",
        "    # Loop through each dataset name\n",
        "    for count, name_data in enumerate(list_ST_name_data):\n",
        "        print(f\"\\nDATA ({count+1}/{len(list_ST_name_data)}): {name_data}\\n\")\n",
        "\n",
        "        # Load the spatial transcriptomics data from the .zarr format\n",
        "        sdata = sd.read_zarr(os.path.join(data_directory_path, f\"{name_data}.zarr\"))\n",
        "\n",
        "        # Extract the list of gene names\n",
        "        gene_name_list = sdata['anucleus'].var['gene_symbols'].values\n",
        "\n",
        "        # Sample the dataset if a subset size is specified\n",
        "        if size_subset is not None:\n",
        "            print(\"Sampling the dataset ...\")\n",
        "            rows_to_keep = list(sdata['anucleus'].obs.sample(n=min(size_subset, len(sdata['anucleus'].obs)), random_state=42).index)\n",
        "        else:\n",
        "            size_subset = len(sdata['anucleus'].obs)\n",
        "            rows_to_keep = list(sdata['anucleus'].obs.sample(n=size_subset, random_state=42).index)\n",
        "\n",
        "        # Extract spatial positions for 'train' cells\n",
        "        cell_id_train = sdata['anucleus'].obs[\"cell_id\"].values\n",
        "        new_spatial_coord = extract_spatial_positions(sdata, cell_id_train)\n",
        "        # Store new spatial coordinates into sdata\n",
        "        sdata['anucleus'].obsm['spatial'] = new_spatial_coord\n",
        "\n",
        "        # Create the gene expression dataset (Y)\n",
        "        print(\"Create gene expression dataset (Y) ...\")\n",
        "        y_subtracted = sdata['anucleus'][rows_to_keep].copy()\n",
        "        # Trick to set all index to same length to avoid problems when saving to h5\n",
        "        y_subtracted.obs.index = ['x' + str(i).zfill(6) for i in y_subtracted.obs.index]\n",
        "\n",
        "        # Save the gene expression data to an H5AD file\n",
        "        y_subtracted.write(os.path.join(adata_save_dir, f'{name_data}.h5ad'))\n",
        "\n",
        "        for index in y_subtracted.obs.index:\n",
        "            if len(index) != len(y_subtracted.obs.index[0]):\n",
        "                warnings.warn(\"indices of y_subtracted.obs should all have the same length to avoid problems when saving to h5\", UserWarning)\n",
        "\n",
        "        # Extract spatial coordinates and barcodes (cell IDs) for the patches\n",
        "        coords_center = y_subtracted.obsm['spatial']\n",
        "        barcodes = np.array(y_subtracted.obs.index)\n",
        "\n",
        "        # Generate and visualize image patches centered around spatial coordinates ({name_data}.h5 file in directory os.path.join(dir_processed_dataset, \"patches\"))\n",
        "        process_and_visualize_image(sdata, patch_save_dir, name_data, coords_center, target_patch_size, barcodes,\n",
        "                                    show_extracted_images=False, vis_width=1000)\n",
        "\n",
        "        # Delete variables that are no longer used\n",
        "        del sdata, y_subtracted\n",
        "        gc.collect()\n",
        "\n",
        "    # Save the gene list to a JSON file\n",
        "    gene_path = os.path.join(dir_processed_dataset, 'var_genes.json')\n",
        "    print(f\"Save gene list in {gene_path}\")\n",
        "    data = {\n",
        "        \"genes\": list(gene_name_list)\n",
        "    }\n",
        "    print(\"Total number of genes:\", len(data[\"genes\"]))\n",
        "\n",
        "    with open(gene_path, \"w\") as f:\n",
        "        json.dump(data, f, indent=4)\n",
        "\n",
        "    print(\"\\nPreprocess dataset DONE:\", \" - \".join(list_ST_name_data), \"\\n\")\n",
        "\n",
        "\n",
        "def preprocess_spatial_transcriptomics_data_test(name_data, sdata, cell_id_list, dir_processed_dataset, target_patch_size=32, vis_width=1000, show_extracted_images=False):\n",
        "    \"\"\"\n",
        "    Test step: Preprocesses spatial transcriptomics data by performing the following steps for the selected ST data:\n",
        "    1. Extract spatial coordinates of the selected cells.\n",
        "    2. Generates and saves patches of images centered on spatial coordinates to HDF5 files (X) into directory 'patches'.\n",
        "\n",
        "    Parameters:\n",
        "    -----------\n",
        "    name_data: str\n",
        "        Name used for saving the dataset.\n",
        "    sdata: SpatialData\n",
        "        A spatial data object containing the image to process ('HE_original') and associated metadata.\n",
        "    cell_id_list : array-like\n",
        "        A list or array of cell IDs to filter the regions.\n",
        "    dir_processed_dataset: str\n",
        "        Path to the directory where processed datasets and outputs will be saved.\n",
        "    target_patch_size: int, optional\n",
        "        Target size of image patches to extract.\n",
        "    vis_width: int, optional\n",
        "        Width of the visualization output for spatial and image patches.\n",
        "    show_extracted_images: bool\n",
        "    \"\"\"\n",
        "\n",
        "    # Creates directories for saving patches ('patches')\n",
        "    patch_save_dir = os.path.join(dir_processed_dataset, \"patches\")\n",
        "    os.makedirs(patch_save_dir, exist_ok=True)\n",
        "\n",
        "    print(\"\\n -- PREPROCESS SPATIAL TRANSCRIPTOMICS DATASET --------------------------------------------\\n\")\n",
        "\n",
        "    # Extract spatial positions for selected cells\n",
        "    new_spatial_coord = extract_spatial_positions(sdata, cell_id_list)\n",
        "\n",
        "    # Spatial coordinates and barcodes (cell IDs) for the patches\n",
        "    coords_center = new_spatial_coord\n",
        "    barcodes = np.array(['x' + str(i).zfill(6) for i in list(cell_id_list)])  # Trick to set all index to same length to avoid problems when saving to h5\n",
        "\n",
        "    # Generate and visualize image patches centered around spatial coordinates ({name_data}.h5 file in directory os.path.join(dir_processed_dataset, \"patches\"))\n",
        "    process_and_visualize_image(sdata, patch_save_dir, name_data, coords_center, target_patch_size, barcodes,\n",
        "                                show_extracted_images=False, vis_width=1000)\n",
        "\n",
        "    print(\"\\nPreprocess dataset DONE\\n\")\n",
        "\n",
        "\n",
        "def create_cross_validation_splits(dir_processed_dataset, n_fold=None):\n",
        "    \"\"\"\n",
        "    Creates cross-validation splits (leave-one-out cv) for spatial transcriptomics data by splitting\n",
        "    samples into training and testing sets and saving them as CSV files.\n",
        "\n",
        "    Example for samples [\"UC1_NI\", \"UC1_I\", \"UC6_NI\"]:\n",
        "      FOLD 0: TRAIN: [\"UC1_NI\", \"UC1_I\"] TEST: [\"UC6_NI\"]\n",
        "      FOLD 1: TRAIN: [\"UC1_NI\", \"UC6_NI\"] TEST: [\"UC1_I\"]\n",
        "      FOLD 2: TRAIN: [\"UC6_NI\", \"UC1_I\"] TEST: [\"UC1_NI\"]\n",
        "\n",
        "    Parameters:\n",
        "    -----------\n",
        "    dir_processed_dataset : str\n",
        "        Path to the directory where processed datasets are saved.\n",
        "    n_fold : int, optional\n",
        "        Number of folds for cross-validation (leave-one-out cv). If None, defaults to number of ST files.\n",
        "    \"\"\"\n",
        "\n",
        "    patches_dir = os.path.join(dir_processed_dataset, \"patches\")\n",
        "    splits_dir = os.path.join(dir_processed_dataset, \"splits\")\n",
        "    os.makedirs(splits_dir, exist_ok=True)\n",
        "\n",
        "    # List all files in the patches directory (these represent individual samples)\n",
        "    patch_files = os.listdir(patches_dir)\n",
        "\n",
        "    # Prepare a list to store information about the samples (patches and gene expression data path)\n",
        "    all_ST = []\n",
        "    # Extra paths by iterating over patch files\n",
        "    for patch_file in patch_files:\n",
        "        if patch_file.endswith('.h5'):\n",
        "            # Extract sample ID from patch file name\n",
        "            sample_id = patch_file.split('.')[0]\n",
        "            # Corresponding gene expression data file (should be in 'adata' directory)\n",
        "            expr_file = os.path.join(\"adata\", f\"{sample_id}.h5ad\")\n",
        "            all_ST.append({\n",
        "                \"sample_id\": sample_id,\n",
        "                \"patches_path\": os.path.join(\"patches\", patch_file),\n",
        "                \"expr_path\": expr_file\n",
        "            })\n",
        "\n",
        "    df_all_ST = pd.DataFrame(all_ST)\n",
        "\n",
        "    # If n_fold is not specified, default to using the number of available samples\n",
        "    if n_fold is None:\n",
        "        n_fold = len(df_all_ST)\n",
        "    # Ensure that the number of folds does not exceed the number of available samples\n",
        "    n_fold = min(n_fold, len(df_all_ST))\n",
        "\n",
        "    print(\"\\n -- CREATE CROSS-VALIDATION SPLITS --------------------------------------------\\n\")\n",
        "\n",
        "    # Generate cross-validation splits (leave-one-out CV)\n",
        "    for i in range(n_fold):\n",
        "        # Select the current sample as the test set (leave-one-out)\n",
        "        test_df = df_all_ST.iloc[[i]]\n",
        "        # Use the remaining samples as the training set\n",
        "        train_df = df_all_ST.drop(i)\n",
        "\n",
        "        print(f\"Index {i}:\")\n",
        "        print(\"Train DataFrame:\")\n",
        "        print(train_df)\n",
        "        print(\"Test DataFrame:\")\n",
        "        print(test_df)\n",
        "\n",
        "        # Save the train and test DataFrames as CSV files in the splits directory\n",
        "        train_filename = f\"train_{i}.csv\"\n",
        "        test_filename = f\"test_{i}.csv\"\n",
        "        train_df.to_csv(os.path.join(splits_dir, train_filename), index=False)\n",
        "        test_df.to_csv(os.path.join(splits_dir, test_filename), index=False)\n",
        "        print(f\"Saved {train_filename} and {test_filename}\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Koaqf4aXwOV0"
      },
      "source": [
        "## Utility functions\n",
        "\n",
        "These utility functions assist with gene expression data processing and model results handling."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {
        "id": "iyslxSHLrV_6"
      },
      "outputs": [],
      "source": [
        "###############\n",
        "# utils fct\n",
        "###############\n",
        "\n",
        "def log1p_normalization(arr):\n",
        "    \"\"\"  Apply log1p normalization to the given array \"\"\"\n",
        "\n",
        "    scale_factor = 100\n",
        "    return np.log1p((arr / np.sum(arr, axis=1, keepdims=True)) * scale_factor)\n",
        "\n",
        "\n",
        "def normalize_adata(adata: sc.AnnData) -> sc.AnnData:\n",
        "    \"\"\"\n",
        "    Normalize and apply log1p transformation to the expression matrix of an AnnData object.\n",
        "    (The function normalizes the gene expression by row)\n",
        "\n",
        "    Parameters:\n",
        "    -----------\n",
        "    adata : sc.AnnData\n",
        "        AnnData object containing gene expression data.\n",
        "    \"\"\"\n",
        "\n",
        "    filtered_adata = adata.copy()\n",
        "    filtered_adata.X = filtered_adata.X.astype(np.float64)\n",
        "    filtered_adata.X = log1p_normalization(filtered_adata.X)\n",
        "\n",
        "    return filtered_adata\n",
        "\n",
        "\n",
        "def load_adata(expr_path, genes=None, barcodes=None, normalize=False):\n",
        "    \"\"\"\n",
        "    Load AnnData object from a given path\n",
        "\n",
        "    Parameters:\n",
        "    -----------\n",
        "    expr_path : str\n",
        "        Path to the .h5ad file containing the AnnData object.\n",
        "    genes : list, optional\n",
        "        List of genes to retain. If None, all genes are kept.\n",
        "    barcodes : list, optional\n",
        "        List of barcodes (cells) to retain. If None, all cells are kept.\n",
        "    normalize : bool, optional\n",
        "        Whether to apply normalization (log1p normalization) to the data.\n",
        "\n",
        "    Returns:\n",
        "    --------\n",
        "    pd.DataFrame\n",
        "        Gene expression data as a DataFrame.\n",
        "    \"\"\"\n",
        "\n",
        "    adata = sc.read_h5ad(expr_path)\n",
        "    if barcodes is not None:\n",
        "        adata = adata[barcodes]\n",
        "    if genes is not None:\n",
        "        adata = adata[:, genes]\n",
        "    if normalize:\n",
        "        adata = normalize_adata(adata)\n",
        "    return adata.to_df()\n",
        "\n",
        "\n",
        "def merge_dict(main_dict, new_dict, value_fn=None):\n",
        "    \"\"\"\n",
        "    Merge new_dict into main_dict. If a key exists in both dicts, the values are appended.\n",
        "    Else, the key-value pair is added.\n",
        "    Expects value to be an array or list - if not, it is converted to a list.\n",
        "    If value_fn is not None, it is applied to each item in each value in new_dict before merging.\n",
        "    Args:\n",
        "        main_dict: main dict\n",
        "        new_dict: new dict\n",
        "        value_fn: function to apply to each item in each value in new_dict before merging\n",
        "    \"\"\"\n",
        "\n",
        "    if value_fn is None:\n",
        "        def value_fn(x): return x\n",
        "\n",
        "    for key, value in new_dict.items():\n",
        "        if not isinstance(value, list):\n",
        "            value = [value]\n",
        "        value = [value_fn(v) for v in value]\n",
        "        if key in main_dict:\n",
        "            main_dict[key] = main_dict[key] + value\n",
        "        else:\n",
        "            main_dict[key] = value\n",
        "\n",
        "    return main_dict\n",
        "\n",
        "\n",
        "def post_collate_fn(batch):\n",
        "    \"\"\" Post collate function to clean up batch \"\"\"\n",
        "\n",
        "    if batch[\"imgs\"].dim() == 5:\n",
        "        assert batch[\"imgs\"].size(0) == 1\n",
        "        batch[\"imgs\"] = batch[\"imgs\"].squeeze(0)\n",
        "\n",
        "    if batch[\"coords\"].dim() == 3:\n",
        "        assert batch[\"coords\"].size(0) == 1\n",
        "        batch[\"coords\"] = batch[\"coords\"].squeeze(0)\n",
        "\n",
        "    return batch\n",
        "\n",
        "\n",
        "def merge_fold_results(arr):\n",
        "    \"\"\" Merges results from multiple cross-validation folds, aggregating Pearson correlation across all folds. \"\"\"\n",
        "    aggr_dict = {}\n",
        "    for dict in arr:\n",
        "        for item in dict['pearson_corrs']:\n",
        "            gene_name = item['name']\n",
        "            correlation = item['pearson_corr']\n",
        "            aggr_dict[gene_name] = aggr_dict.get(gene_name, []) + [correlation]\n",
        "\n",
        "    aggr_results = []\n",
        "    all_corrs = []\n",
        "    for key, value in aggr_dict.items():\n",
        "        aggr_results.append({\n",
        "            \"name\": key,\n",
        "            \"pearson_corrs\": value,\n",
        "            \"mean\": np.mean(value),\n",
        "            \"std\": np.std(value)\n",
        "        })\n",
        "        all_corrs += value\n",
        "\n",
        "    mean_per_split = [d['pearson_mean'] for d in arr]\n",
        "\n",
        "    return {\n",
        "        \"pearson_corrs\": aggr_results,\n",
        "        \"pearson_mean\": np.mean(mean_per_split),\n",
        "        \"pearson_std\": np.std(mean_per_split),\n",
        "        \"mean_per_split\": mean_per_split\n",
        "    }"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "0tBRiWNowz6b"
      },
      "source": [
        "## Encoder functions\n",
        "\n",
        "Provide the structure for building and utilizing pre-trained models specifically designed for extracting features from images."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {
        "id": "MBn6-RVJ4THe"
      },
      "outputs": [],
      "source": [
        "##########\n",
        "# Encoder functions\n",
        "##########\n",
        "\n",
        "class InferenceEncoder(torch.nn.Module):\n",
        "    \"\"\"\n",
        "    Abstract base class for building inference encoders.\n",
        "\n",
        "    Attributes:\n",
        "    -----------\n",
        "    weights_path : str or None\n",
        "        Path to the model weights (optional).\n",
        "    model : torch.nn.Module\n",
        "        The model architecture.\n",
        "    eval_transforms : callable\n",
        "        Evaluation transformations applied to the input images.\n",
        "    precision : torch.dtype\n",
        "        The data type of the model's parameters and inputs.\n",
        "    \"\"\"\n",
        "\n",
        "    def __init__(self, weights_path=None, **build_kwargs):\n",
        "        super(InferenceEncoder, self).__init__()\n",
        "\n",
        "        self.weights_path = weights_path\n",
        "        self.model, self.eval_transforms, self.precision = self._build(weights_path, **build_kwargs)\n",
        "\n",
        "    def forward(self, x):\n",
        "        z = self.model(x)\n",
        "        return z\n",
        "\n",
        "    @abstractmethod\n",
        "    def _build(self, **build_kwargs):\n",
        "        pass\n",
        "\n",
        "from torchvision import transforms\n",
        "import torchvision.transforms.functional as TF\n",
        "\n",
        "def get_eval_transforms(mean, std):\n",
        "    \"\"\"\n",
        "    Creates the evaluation transformations for preprocessing images. This includes\n",
        "    converting the images to tensor format and normalizing them with given mean and std.\n",
        "\n",
        "    Parameters:\n",
        "    -----------\n",
        "    mean : list\n",
        "        The mean values used for normalization.\n",
        "    std : list\n",
        "        The standard deviation values used for normalization.\n",
        "\n",
        "    Returns:\n",
        "    --------\n",
        "    transforms.Compose\n",
        "        A composed transformation function that applies the transformations in sequence.\n",
        "    \"\"\"\n",
        "    trsforms = []\n",
        "\n",
        "    # Convert image to tensor\n",
        "    trsforms.append(lambda img: TF.to_tensor(img))\n",
        "\n",
        "    if mean is not None and std is not None:\n",
        "        # Normalize the image\n",
        "        trsforms.append(lambda img: TF.normalize(img, mean, std))\n",
        "\n",
        "    return transforms.Compose(trsforms)\n",
        "\n",
        "class ResNet50InferenceEncoder(InferenceEncoder):\n",
        "    \"\"\"\n",
        "    A specific implementation of the InferenceEncoder class for ResNet50.\n",
        "    This encoder is used to extract features from images using a pretrained ResNet50 model.\n",
        "    \"\"\"\n",
        "\n",
        "    def _build(\n",
        "        self,\n",
        "        weights_root=\"resnet50.tv_in1k\",\n",
        "        timm_kwargs={\"features_only\": True, \"out_indices\": [3], \"num_classes\": 0},\n",
        "        pool=True\n",
        "    ):\n",
        "        \"\"\"\n",
        "        Build the ResNet50 model and load its weights. It supports both pretrained models\n",
        "        from the internet and pretrained models from a given weights path (offline).\n",
        "\n",
        "        Parameters:\n",
        "        -----------\n",
        "        weights_root : str\n",
        "            Path to pretrained model weights. Defaults to \"resnet50.tv_in1k\" (if online).\n",
        "        timm_kwargs : dict\n",
        "            Additional arguments for creating the ResNet50 model via the timm library.\n",
        "        pool : bool\n",
        "            Whether to apply adaptive average pooling to the output of the model. Defaults to True.\n",
        "\n",
        "        Returns:\n",
        "        --------\n",
        "        tuple\n",
        "            A tuple containing the ResNet50 model, the evaluation transformations, and the precision type.\n",
        "        \"\"\"\n",
        "\n",
        "        if weights_root == \"resnet50.tv_in1k\":\n",
        "            pretrained = True\n",
        "            print(\"Load pretrained Resnet50 from internet\")\n",
        "        else:\n",
        "            pretrained = False\n",
        "            print(f\"Load pretrained Resnet50 offline from weights path: {weights_root}\")\n",
        "\n",
        "        # Build the model using the timm library\n",
        "        model = timm.create_model(\"resnet50.tv_in1k\", pretrained=pretrained, **timm_kwargs)\n",
        "\n",
        "        # If not using a pretrained model, load weights from the specified path\n",
        "        if not pretrained and os.path.exists(weights_root):\n",
        "            # Load the weights\n",
        "            checkpoint = torch.load(weights_root, map_location='cpu', weights_only=True)  # or 'cuda' if using GPU\n",
        "\n",
        "            # Remove the classifier layers from the checkpoint\n",
        "            model_state_dict = model.state_dict()\n",
        "            checkpoint = {k: v for k, v in checkpoint.items() if k in model_state_dict}\n",
        "\n",
        "            # Load the weights into the model\n",
        "            model.load_state_dict(checkpoint, strict=False)\n",
        "        elif not pretrained:\n",
        "            # Issue a warning if the weights file is missing\n",
        "            print(f\"\\n!!! WARNING: The specified weights file '{weights_root}' does not exist. The model will be initialized with random weights.\\n\")\n",
        "\n",
        "        imagenet_mean, imagenet_std = [0.485, 0.456, 0.406], [0.229, 0.224, 0.225]\n",
        "        eval_transform = get_eval_transforms(imagenet_mean, imagenet_std)\n",
        "        precision = torch.float32\n",
        "        if pool:\n",
        "            self.pool = torch.nn.AdaptiveAvgPool2d(1)\n",
        "        else:\n",
        "            self.pool = None\n",
        "\n",
        "        return model, eval_transform, precision\n",
        "\n",
        "    def forward(self, x):\n",
        "        out = self.forward_features(x)\n",
        "        if self.pool:\n",
        "            out = self.pool(out).squeeze(-1).squeeze(-1)\n",
        "        return out\n",
        "\n",
        "    def forward_features(self, x):\n",
        "        out = self.model(x)\n",
        "        if isinstance(out, list):\n",
        "            assert len(out) == 1\n",
        "            out = out[0]\n",
        "        return out\n",
        "\n",
        "def inf_encoder_factory(enc_name):\n",
        "    \"\"\"\n",
        "    Factory function to instantiate an encoder based on the specified name.\n",
        "\n",
        "    Parameters:\n",
        "    -----------\n",
        "    enc_name : str\n",
        "        The name of the encoder model to instantiate (e.g., 'resnet50').\n",
        "\n",
        "    Returns:\n",
        "    --------\n",
        "    class\n",
        "        The encoder class corresponding to the specified encoder name.\n",
        "    \"\"\"\n",
        "\n",
        "    if enc_name == 'resnet50':\n",
        "        return ResNet50InferenceEncoder\n",
        "\n",
        "    raise ValueError(f\"Unknown encoder name {enc_name}\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "f_gewG-TqZL_"
      },
      "source": [
        "## Torch Dataset & Embeddings\n",
        "\n",
        "This section defines a custom torch dataset for handling spatial transcriptomics data and generating embeddings for image patches.\n",
        "\n",
        "- **`H5Dataset`**: A torch `Dataset` class for loading spatial transcriptomics data from an HDF5 file. It loads image patches, their associated barcodes/cells and coordinates in batches to efficiently handle large datasets.\n",
        "- **`generate_embeddings`**: A utility for generating embeddings from images  and saving them to an HDF5 file. It handles creating a `DataLoader`, running the model in evaluation mode and saving embeddings to disk."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "metadata": {
        "id": "g3pPQPc6EKEq"
      },
      "outputs": [],
      "source": [
        "#########\n",
        "# Torch Dataset & Embeddings\n",
        "#########\n",
        "from torch.utils.data import Dataset\n",
        "\n",
        "\n",
        "class H5Dataset(Dataset):\n",
        "    \"\"\"\n",
        "    Dataset to read ST + H&E from an HDF5 (.h5) file\n",
        "    The dataset loads images and their associated barcodes/cells and coordinates in chunks for efficient data handling.\n",
        "\n",
        "    Attributes:\n",
        "    -----------\n",
        "    h5_path : str\n",
        "        Path to the HDF5 file containing the images, barcodes, and coordinates.\n",
        "    img_transform : callable, optional\n",
        "        A transformation function to apply to the images. Defaults to None.\n",
        "    chunk_size : int, optional\n",
        "        Number of items to load per batch. Defaults to 1000.\n",
        "    n_chunks : int\n",
        "        The total number of chunks, calculated based on the size of the 'barcode' array.\n",
        "    \"\"\"\n",
        "\n",
        "    def __init__(self, h5_path, img_transform=None, chunk_size=1000):\n",
        "        self.h5_path = h5_path\n",
        "        self.img_transform = img_transform\n",
        "        self.chunk_size = chunk_size\n",
        "\n",
        "        with h5py.File(h5_path, 'r') as f:\n",
        "            self.n_chunks = int(np.ceil(len(f['barcode']) / chunk_size))\n",
        "\n",
        "    def __len__(self):\n",
        "        return self.n_chunks\n",
        "\n",
        "    def __getitem__(self, idx):\n",
        "        \"\"\"\n",
        "        Fetches a batch of data (images, barcodes, and coordinates) from the HDF5 file.\n",
        "\n",
        "        Parameters:\n",
        "        -----------\n",
        "        idx : int\n",
        "            The index of the chunk to fetch.\n",
        "\n",
        "        Returns:\n",
        "        --------\n",
        "        dict\n",
        "            A dictionary containing the images, barcodes, and coordinates for the specified chunk.\n",
        "        \"\"\"\n",
        "\n",
        "        start_idx = idx * self.chunk_size\n",
        "        end_idx = (idx + 1) * self.chunk_size\n",
        "        # Open the HDF5 file and load the specific chunk of data\n",
        "        with h5py.File(self.h5_path, 'r') as f:\n",
        "            imgs = f['img'][start_idx:end_idx]\n",
        "            barcodes = f['barcode'][start_idx:end_idx].flatten().tolist()\n",
        "            coords = f['coords'][start_idx:end_idx]\n",
        "\n",
        "        # Apply image transformations if any (e.g. to Torch and normalization)\n",
        "        if self.img_transform:\n",
        "            imgs = torch.stack([self.img_transform(img) for img in imgs])\n",
        "\n",
        "        return {'imgs': imgs, 'barcodes': barcodes, 'coords': coords}\n",
        "\n",
        "\n",
        "def embed_tiles(dataloader, model: torch.nn.Module, embedding_save_path: str, device: str, precision):\n",
        "    \"\"\"\n",
        "    Extracts embeddings from image tiles using the specified model and saves them to an HDF5 file.\n",
        "\n",
        "    Parameters:\n",
        "    -----------\n",
        "    dataloader : torch.utils.data.DataLoader\n",
        "        DataLoader providing the batches of image tiles.\n",
        "    model : torch.nn.Module\n",
        "        The model used to generate embeddings from the tiles.\n",
        "    embedding_save_path : str\n",
        "        Path where the generated embeddings will be saved.\n",
        "    device : str\n",
        "        The device to run the model on (e.g., 'cuda' or 'cpu').\n",
        "    precision : torch.dtype\n",
        "        The precision (data type) to use for inference (e.g., float16 for mixed precision).\n",
        "    \"\"\"\n",
        "\n",
        "    model.eval()\n",
        "    # Iterate over the batches in the DataLoader\n",
        "    for batch_idx, batch in tqdm(enumerate(dataloader), total=len(dataloader)):\n",
        "        batch = post_collate_fn(batch)\n",
        "        imgs = batch['imgs'].to(device).float()\n",
        "        # Apply model on images\n",
        "        with torch.inference_mode():\n",
        "            if torch.cuda.is_available():  # Use mixed precision only if CUDA is available\n",
        "                with torch.amp.autocast('cuda', dtype=precision):\n",
        "                    embeddings = model(imgs)\n",
        "            else:  # No mixed precision on CPU\n",
        "                embeddings = model(imgs)\n",
        "\n",
        "        # Set mode to 'w' for the first batch, 'a' for appending subsequent batches\n",
        "        mode = 'w' if batch_idx == 0 else 'a'\n",
        "\n",
        "        # Create a dictionary with embeddings and other relevant data to save\n",
        "        asset_dict = {'embeddings': embeddings.cpu().numpy()}\n",
        "        asset_dict.update({key: np.array(val) for key, val in batch.items() if key != 'imgs'})\n",
        "\n",
        "        # Save the embeddings to the HDF5 file\n",
        "        save_hdf5(embedding_save_path, asset_dict=asset_dict, mode=mode)\n",
        "\n",
        "    return embedding_save_path\n",
        "\n",
        "\n",
        "def generate_embeddings(embed_path, encoder, device, tile_h5_path, batch_size, num_workers, overwrite=False):\n",
        "    \"\"\"\n",
        "    Generate embeddings for images and save to a specified path.\n",
        "\n",
        "    Parameters:\n",
        "    -----------\n",
        "    embed_path : str\n",
        "        Path to save the embeddings.\n",
        "    encoder : torch.nn.Module\n",
        "        The encoder model for generating embeddings.\n",
        "    device : torch.device\n",
        "        Device to use for computation (e.g., 'cuda' or 'cpu').\n",
        "    tile_h5_path : str\n",
        "        Path to the HDF5 file containing images.\n",
        "    batch_size : int\n",
        "        Batch size for the DataLoader.\n",
        "    num_workers : int\n",
        "        Number of worker threads for data loading.\n",
        "    overwrite : bool, optional\n",
        "        If True, overwrite existing embeddings. Default is False.\n",
        "    \"\"\"\n",
        "\n",
        "    # If the embeddings file doesn't exist or overwrite is True, proceed to generate embeddings\n",
        "    if not os.path.isfile(embed_path) or overwrite:\n",
        "        print(f\"Generating embeddings for {embed_path} ...\")\n",
        "\n",
        "        # Set encoder to evaluation mode and move it to the device\n",
        "        encoder.eval()\n",
        "        encoder.to(device)\n",
        "\n",
        "        # Create dataset and dataloader for tiles\n",
        "        tile_dataset = H5Dataset(tile_h5_path, chunk_size=batch_size, img_transform=encoder.eval_transforms)\n",
        "        tile_dataloader = torch.utils.data.DataLoader(\n",
        "            tile_dataset,\n",
        "            batch_size=1,\n",
        "            shuffle=False,\n",
        "            num_workers=num_workers\n",
        "        )\n",
        "\n",
        "        # Generate and save embeddings\n",
        "        embed_tiles(tile_dataloader, encoder, embed_path, device, encoder.precision)\n",
        "    else:\n",
        "        print(f\"Skipping embedding {os.path.basename(embed_path)} as it already exists\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "m95S5XlMq6gL"
      },
      "source": [
        "## Regression Functions\n",
        "\n",
        "This section defines functions for training a regression model and evaluating its performance.\n",
        "\n",
        "* **`compute_metrics`**: Computes evaluation metrics, including L2 error, R² score and Pearson correlation for each target/gene.\n",
        "* **`train_test_reg`**: Trains a regression model (default: Ridge regression) using the provided embedding vector train data and evaluates it on test data."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {
        "id": "_imPcy_bG89u"
      },
      "outputs": [],
      "source": [
        "##########\n",
        "# Regression functions\n",
        "##########\n",
        "\n",
        "def compute_metrics(y_test, preds_all, genes=None):\n",
        "    \"\"\"\n",
        "    Computes metrics L2 errors R2 scores and Pearson correlations for each target/gene.\n",
        "\n",
        "    :param y_test: Ground truth values (numpy array of shape [n_samples, n_targets]).\n",
        "    :param preds_all: Predictions for all targets (numpy array of shape [n_samples, n_targets]).\n",
        "    :param genes: Optional list of gene names corresponding to targets.\n",
        "    :return: A dictionary containing metrics.\n",
        "    \"\"\"\n",
        "\n",
        "    errors = []\n",
        "    r2_scores = []\n",
        "    pearson_corrs = []\n",
        "    pearson_genes = []\n",
        "\n",
        "    for i, target in enumerate(range(y_test.shape[1])):\n",
        "        preds = preds_all[:, target]\n",
        "        target_vals = y_test[:, target]\n",
        "\n",
        "        # Compute L2 error\n",
        "        l2_error = float(np.mean((preds - target_vals) ** 2))\n",
        "\n",
        "        # Compute R2 score\n",
        "        total_variance = np.sum((target_vals - np.mean(target_vals)) ** 2)\n",
        "        if total_variance == 0:\n",
        "            r2_score = 0.0\n",
        "        else:\n",
        "            r2_score = float(1 - np.sum((target_vals - preds) ** 2) / total_variance)\n",
        "\n",
        "        # Compute Pearson correlation\n",
        "        with warnings.catch_warnings():\n",
        "            warnings.filterwarnings(\"error\", category=ConstantInputWarning)\n",
        "            try:\n",
        "                pearson_corr, _ = pearsonr(target_vals, preds)\n",
        "                pearson_corr = pearson_corr if not np.isnan(pearson_corr) else 0.0\n",
        "            except ConstantInputWarning:\n",
        "                pearson_corr = 0.0\n",
        "\n",
        "        errors.append(l2_error)\n",
        "        r2_scores.append(r2_score)\n",
        "        pearson_corrs.append(pearson_corr)\n",
        "\n",
        "        # Record gene-specific Pearson correlation\n",
        "        if genes is not None:\n",
        "            pearson_genes.append({\n",
        "                'name': genes[i],\n",
        "                'pearson_corr': pearson_corr\n",
        "            })\n",
        "\n",
        "    # Compile results\n",
        "    results = {\n",
        "        'pearson_mean': float(np.mean(pearson_corrs)),\n",
        "        'l2_errors_mean': float(np.mean(errors)),\n",
        "        'r2_scores_mean': float(np.mean(r2_scores)),\n",
        "        'l2_errors': list(errors),\n",
        "        'r2_scores': list(r2_scores),\n",
        "        'pearson_corrs': pearson_genes if genes is not None else list(pearson_corrs),\n",
        "        'pearson_std': float(np.std(pearson_corrs)),\n",
        "        'l2_error_q1': float(np.percentile(errors, 25)),\n",
        "        'l2_error_q2': float(np.median(errors)),\n",
        "        'l2_error_q3': float(np.percentile(errors, 75)),\n",
        "        'r2_score_q1': float(np.percentile(r2_scores, 25)),\n",
        "        'r2_score_q2': float(np.median(r2_scores)),\n",
        "        'r2_score_q3': float(np.percentile(r2_scores, 75)),\n",
        "    }\n",
        "\n",
        "    return results\n",
        "\n",
        "\n",
        "def train_test_reg(X_train, X_test, y_train, y_test,\n",
        "                   max_iter=1000, random_state=0, genes=None, alpha=None, method='ridge'):\n",
        "    \"\"\" Train a regression model and evaluate its performance on test data \"\"\"\n",
        "\n",
        "    if method == 'ridge':\n",
        "        # If alpha is not provided, compute it based on the input dimensions\n",
        "        alpha = 100 / (X_train.shape[1] * y_train.shape[1])\n",
        "\n",
        "        print(f\"Ridge: using alpha: {alpha}\")\n",
        "        # Initialize Ridge regression model\n",
        "        reg = Ridge(solver='lsqr',\n",
        "                    alpha=alpha,\n",
        "                    random_state=random_state,\n",
        "                    fit_intercept=False,\n",
        "                    max_iter=max_iter)\n",
        "        # Fit the model on the training data\n",
        "        reg.fit(X_train, y_train)\n",
        "\n",
        "        # Make predictions on the test data\n",
        "        preds_all = reg.predict(X_test)\n",
        "\n",
        "    # You can instantiate other regression models...\n",
        "\n",
        "    # Compute the evaluation metrics using the test data and predictions\n",
        "    results = compute_metrics(y_test, preds_all, genes=genes)\n",
        "\n",
        "    return reg, results"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "---kKu3RrfH4"
      },
      "source": [
        "## Training Functions\n",
        "\n",
        "This section defines functions for training models and saving results.\n",
        "\n",
        "- **`predict_single_split`**: Performs predictions for a single train-test split. It handles the embedding generation, loading of gene expression data, model training and saving the results. The trained model is saved to a specified directory.\n",
        "- **`predict_folds`**: Performs predictions for all train-test folds, iterating over the splits and calling `predict_single_split` for each fold. It saves the results for all folds, including a summary of Pearson correlations, L2 errors and R² scores.\n",
        "- **`run_training`**: Executes the full training process by calling `predict_folds` and saving the encoder performance results."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "metadata": {
        "id": "YI7ySRvP2-qs"
      },
      "outputs": [],
      "source": [
        "##########\n",
        "# Training functions\n",
        "##########\n",
        "\n",
        "def predict_single_split(fold, train_split, test_split, args, save_dir, model_name, device, bench_data_root):\n",
        "    \"\"\" Predict a single split for a single model \"\"\"\n",
        "\n",
        "    # Ensure paths to train and test split files are correct\n",
        "    if not os.path.isfile(train_split):\n",
        "        train_split = os.path.join(bench_data_root, 'splits', train_split)\n",
        "\n",
        "    if not os.path.isfile(test_split):\n",
        "        test_split = os.path.join(bench_data_root, 'splits', test_split)\n",
        "\n",
        "    # Read train and test split CSV files\n",
        "    train_df = pd.read_csv(train_split)\n",
        "    test_df = pd.read_csv(test_split)\n",
        "\n",
        "    # Directory to save embedding results\n",
        "    embedding_dir = args.embed_dataroot\n",
        "    os.makedirs(embedding_dir, exist_ok=True)\n",
        "\n",
        "    # Embedding process\n",
        "    print(\"\\n--EMBEDDING--\\n\")\n",
        "    print(f\"Embedding tiles using {model_name} encoder\")\n",
        "    encoder: InferenceEncoder = inf_encoder_factory(model_name)(args.weights_root)\n",
        "\n",
        "    # Loop over train and test splits to generate embeddings\n",
        "    for split in [train_df, test_df]:\n",
        "        for i in range(len(split)):\n",
        "            sample_id = split.iloc[i]['sample_id']\n",
        "            tile_h5_path = os.path.join(bench_data_root, split.iloc[i]['patches_path'])\n",
        "            assert os.path.isfile(tile_h5_path)\n",
        "            embed_path = os.path.join(embedding_dir, f'{sample_id}.h5')\n",
        "            print(f\"\\nGENERATE EMBEDDING - {sample_id}\\n\")\n",
        "            generate_embeddings(embed_path, encoder, device, tile_h5_path, args.batch_size, args.num_workers, overwrite=args.overwrite)\n",
        "\n",
        "    # Initialize dictionary to hold data for both splits\n",
        "    all_split_assets = {}\n",
        "\n",
        "    gene_list = args.gene_list\n",
        "\n",
        "    # print(f'using gene_list {gene_list}')\n",
        "    # Load gene list for expression data\n",
        "    with open(os.path.join(bench_data_root, gene_list), 'r') as f:\n",
        "        genes = json.load(f)['genes']\n",
        "\n",
        "    # Process train and test splits\n",
        "    for split_key, split in zip(['train', 'test'], [train_df, test_df]):\n",
        "        split_assets = {}\n",
        "        for i in tqdm(range(len(split))):\n",
        "            # Get sample ID, embedding path and gene expression path\n",
        "            sample_id = split.iloc[i]['sample_id']\n",
        "            embed_path = os.path.join(embedding_dir, f'{sample_id}.h5')\n",
        "            expr_path = os.path.join(bench_data_root, split.iloc[i]['expr_path'])\n",
        "\n",
        "            # Read embedding data and gene expression data\n",
        "            assets, _ = read_assets_from_h5(embed_path)\n",
        "            barcodes = assets['barcodes'].flatten().astype(str).tolist()\n",
        "            adata = load_adata(expr_path, genes=genes, barcodes=barcodes, normalize=args.normalize)\n",
        "            assets['adata'] = adata.values\n",
        "\n",
        "            # Merge assets for the split\n",
        "            split_assets = merge_dict(split_assets, assets)\n",
        "\n",
        "        # Concatenate all data in the split\n",
        "        for key, val in split_assets.items():\n",
        "            split_assets[key] = np.concatenate(val, axis=0)\n",
        "\n",
        "        all_split_assets[split_key] = split_assets\n",
        "        print(f\"Loaded {split_key} split with {len(split_assets['embeddings'])} samples: {split_assets['embeddings'].shape}\")\n",
        "\n",
        "    # Assign data to X and y variables for training and testing\n",
        "    X_train, y_train = all_split_assets['train']['embeddings'], all_split_assets['train']['adata']\n",
        "    X_test, y_test = all_split_assets['test']['embeddings'], all_split_assets['test']['adata']\n",
        "\n",
        "    print(\"\\n--REGRESSION--\\n\")\n",
        "    # Perform regression using the specified method\n",
        "    reg, probe_results = train_test_reg(X_train, X_test, y_train, y_test, random_state=args.seed, genes=genes, method=args.method)\n",
        "\n",
        "    # Save the trained regression model\n",
        "    model_path = os.path.join(save_dir, f'model.pkl')\n",
        "    joblib.dump(reg, model_path)\n",
        "    print(f\"Model saved in '{model_path}'\")\n",
        "\n",
        "    # Summarize results for the current fold\n",
        "    print(f\"\\n--FOLD {fold} RESULTS--\\n\")\n",
        "    probe_summary = {}\n",
        "    probe_summary.update({'n_train': len(y_train), 'n_test': len(y_test)})\n",
        "    probe_summary.update({key: val for key, val in probe_results.items()})\n",
        "    keys_to_print = [\"n_train\", \"n_test\", \"pearson_mean\", \"l2_errors_mean\", \"r2_scores_mean\", \"l2_error_q1\", \"l2_error_q2\", \"l2_error_q3\"]\n",
        "\n",
        "    filtered_summary = {\n",
        "        key: round(probe_summary[key], 4)\n",
        "        for key in keys_to_print\n",
        "        if key in probe_summary\n",
        "    }\n",
        "    print(filtered_summary)\n",
        "\n",
        "    with open(os.path.join(save_dir, f'results.json'), 'w') as f:\n",
        "        json.dump(probe_results, f, sort_keys=True, indent=4)\n",
        "\n",
        "    with open(os.path.join(save_dir, f'summary.json'), 'w') as f:\n",
        "        json.dump(probe_summary, f, sort_keys=True, indent=4)\n",
        "\n",
        "    return probe_results\n",
        "\n",
        "\n",
        "def predict_folds(args, exp_save_dir, model_name, device, bench_data_root):\n",
        "    \"\"\" Predict all folds for a given model \"\"\"\n",
        "\n",
        "    # Define the directory for splits\n",
        "    split_dir = os.path.join(bench_data_root, 'splits')\n",
        "    if not os.path.exists(split_dir):\n",
        "        raise FileNotFoundError(f\"{split_dir} doesn't exist, make sure that you specified the splits directory\")\n",
        "\n",
        "    splits = os.listdir(split_dir)\n",
        "    if len(splits) == 0:\n",
        "        raise FileNotFoundError(f\"{split_dir} is empty, make sure that you specified train and test files\")\n",
        "\n",
        "    n_splits = len(splits) // 2\n",
        "\n",
        "    # Save training configuration to JSON\n",
        "    with open(os.path.join(exp_save_dir, 'config.json'), 'w') as f:\n",
        "        json.dump(vars(args), f, sort_keys=True, indent=4)\n",
        "\n",
        "    # Loop through each split and perform predictions\n",
        "    libprobe_results_arr = []\n",
        "    for i in range(n_splits):\n",
        "        print(f\"\\n--FOLD {i}--\\n\")\n",
        "        train_split = os.path.join(split_dir, f'train_{i}.csv')\n",
        "        test_split = os.path.join(split_dir, f'test_{i}.csv')\n",
        "        kfold_save_dir = os.path.join(exp_save_dir, f'split{i}')\n",
        "        os.makedirs(kfold_save_dir, exist_ok=True)\n",
        "\n",
        "        # Predict using the current fold\n",
        "        linprobe_results = predict_single_split(i, train_split, test_split, args, kfold_save_dir, model_name, device=device, bench_data_root=bench_data_root)\n",
        "        libprobe_results_arr.append(linprobe_results)\n",
        "\n",
        "    # Merge and save k-fold results\n",
        "    kfold_results = merge_fold_results(libprobe_results_arr)\n",
        "    with open(os.path.join(exp_save_dir, f'results_kfold.json'), 'w') as f:\n",
        "        p_corrs = kfold_results['pearson_corrs']\n",
        "        p_corrs = sorted(p_corrs, key=itemgetter('mean'), reverse=True)\n",
        "        kfold_results['pearson_corrs'] = p_corrs\n",
        "        json.dump(kfold_results, f, sort_keys=True, indent=4)\n",
        "\n",
        "    return kfold_results\n",
        "\n",
        "\n",
        "def run_training(args):\n",
        "    \"\"\" Training function: Execute predict_folds for processed train dataset with the specified encoder and dump the results in a nested directory structure \"\"\"\n",
        "\n",
        "    print(\"\\n-- RUN TRAINING ---------------------------------------------------------------\\n\")\n",
        "\n",
        "    print(f'run parameters {args}')\n",
        "    # Set device to GPU if available\n",
        "    device = torch.device(\"cuda\" if torch.cuda.is_available() else \"cpu\")\n",
        "\n",
        "    # Create directory for results\n",
        "    save_dir = args.results_dir\n",
        "    os.makedirs(save_dir, exist_ok=True)\n",
        "\n",
        "    # Perform predictions for all folds\n",
        "    enc_results = predict_folds(args, save_dir, model_name=args.encoder, device=device, bench_data_root=args.dir_dataset)\n",
        "\n",
        "    # Store and save encoder performance results\n",
        "    enc_perfs = {\n",
        "        'encoder_name': args.encoder,\n",
        "        'pearson_mean': round(enc_results['pearson_mean'], 4),\n",
        "        'pearson_std': round(enc_results['pearson_std'], 4),\n",
        "    }\n",
        "\n",
        "    with open(os.path.join(save_dir, 'dataset_results.json'), 'w') as f:\n",
        "        json.dump(enc_perfs, f, sort_keys=True, indent=4)\n",
        "\n",
        "    print(\"\\n-- TRAINING DONE ---------------------------------------------------------------\\n\")\n",
        "\n",
        "    print(\"\\n-- Leave-one-out CV performance ---------------------------------------------------------------\")\n",
        "    print(enc_perfs)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "mLCzt4NwsqV6"
      },
      "source": [
        "## Train function\n",
        "\n",
        "The `train` function automates the training of models for spatial transcriptomics data.\n",
        "\n",
        "1. Prepare the input data by preprocessing the spatial transcriptomics datasets (image patches (X) and gene expression (Y)).\n",
        "2. Create leave-one-out cross-validation splits for training and testing.\n",
        "3. Use the specified encoder model (ResNet50) and a regression method (ridge regression) to train the model\n",
        "4. Save the results and the trained model in the `.resources/` directories for later inference.\n",
        "\n",
        "![train_architecture](https://raw.githubusercontent.com/crunchdao/competitions/refs/heads/master/competitions/broad-1/quickstarters/resnet50-plus-ridge/images/train_architecture.png)\n",
        "\n",
        "`Y` is log1p-normalized with scale factor 100."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 12,
      "metadata": {
        "id": "PVnex5Qo-pUl"
      },
      "outputs": [],
      "source": [
        "# In the training function, users build and train the model to make inferences on the test data.\n",
        "# Your train models must be stored in the `model_directory_path`.\n",
        "def train(\n",
        "    data_directory_path: str, # Path to the input data directory\n",
        "    model_directory_path: str # Path to save the trained model and results\n",
        "):\n",
        "    print(\"\\n-- TRAINING ---------------------------------------------------------------\\n\")\n",
        "\n",
        "    os.makedirs(model_directory_path, exist_ok=True)\n",
        "\n",
        "    # Directory to store the processed train dataset (temporary storage)\n",
        "    dir_processed_dataset = os.path.join(\"/tmp\", f\"processed_dataset\")\n",
        "    os.makedirs(dir_processed_dataset, exist_ok=True)\n",
        "\n",
        "    # Directory to store models and results\n",
        "    dir_models_and_results = os.path.join(model_directory_path, f\"ST_pred_results\")\n",
        "    os.makedirs(dir_models_and_results, exist_ok=True)\n",
        "\n",
        "    # List of datasets for Spatial Transcriptomics (ST) training\n",
        "    # The DC1 sample is deliberately absent because it does not contain spatial transcriptomic data.\n",
        "    list_ST_name_data = [\"UC1_NI\", \"UC1_I\", \"UC6_NI\", \"UC6_I\", \"UC7_I\", \"UC9_I\", \"DC5\"]\n",
        "\n",
        "    # Training parameters\n",
        "    args_dict = {\n",
        "        # Parameters for data preprocessing\n",
        "        \"size_subset\": 10000,  # Sampling 10,000 smaller images from each H&E image\n",
        "        \"target_patch_size\": 32,  # Target size of cell patches (sub-region of an image) for the data\n",
        "\n",
        "        \"show_extracted_images\": False,  # (Only for visualization) Whether to visualize all the extracted patches for the first ST data\n",
        "        \"vis_width\": 1000,  # (Only for visualization) Width of the above visualization\n",
        "\n",
        "        # Generic training settings\n",
        "        \"seed\": 1,  # Random seed for reproducibility\n",
        "        \"overwrite\": False,  # Whether to overwrite the existing embedding files (Set to True if you have just changed some data preprocessing parameters)\n",
        "        \"dir_dataset\": dir_processed_dataset,  # Path to save the processed train dataset\n",
        "        \"embed_dataroot\": os.path.join(dir_processed_dataset, f\"ST_data_emb\"),  # Path for embedding data\n",
        "        \"results_dir\": dir_models_and_results,  # Directory to save results\n",
        "        \"n_fold\": 2,  # Number of folds for leave-one-out cross-validation\n",
        "\n",
        "        # Encoder settings\n",
        "        \"batch_size\": 128,  # Batch size for training\n",
        "        \"num_workers\": 0,  # Number of workers for loading data (set 0 for submission on CrunchDAO plaform)\n",
        "\n",
        "        # Train settings\n",
        "        \"gene_list\": \"var_genes.json\",  # Path to save the list of genes\n",
        "        \"method\": \"ridge\",  # Regression method to use\n",
        "        \"alpha\": None,  # Regularization parameter for ridge regression\n",
        "        \"normalize\": False,  # Whether to normalize the data (sdata[\"anucleus\"].X is already normalized with scale factor 100)\n",
        "        \"encoder\": \"resnet50\",  # Encoder model to use (e.g., ResNet50)\n",
        "        \"weights_root\": os.path.join(model_directory_path, f\"pytorch_model.bin\")  # Path to the pre-trained model weights\n",
        "    }\n",
        "\n",
        "    # Convert the dictionary to a simple namespace for easy access to attributes\n",
        "    args = SimpleNamespace(**args_dict)\n",
        "\n",
        "    # Preprocess the spatial transcriptomics data for each sample in list_ST_name_data -> all preprocessed data saved in dir_processed_dataset\n",
        "    preprocess_spatial_transcriptomics_data_train(list_ST_name_data, data_directory_path, dir_processed_dataset,\n",
        "                                                  args.size_subset, args.target_patch_size, args.vis_width, args.show_extracted_images)\n",
        "\n",
        "    # Create leave-one-out cross-validation splits for training and testing in csv files\n",
        "    create_cross_validation_splits(dir_processed_dataset, n_fold=args.n_fold)\n",
        "\n",
        "    # Run the training process using the parameters set above\n",
        "    run_training(args)\n",
        "\n",
        "    # View all results in directory `args.results_dir`"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "-CokrevgcHdm"
      },
      "source": [
        "All regression models and metric results are saved in `./resources/ST_pred_results` -- check it out!\n",
        "\n",
        "Preprocessed datasets are saved in `/tmp/processed_dataset` (temporary storage - not needed for inference).\n",
        "\n",
        "Official challenge metric: L2 mean error (`l2_errors_mean`)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 13,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 478
        },
        "id": "mre9PhC9AQMl",
        "outputId": "da20f729-3d30-467c-9272-f75e0f3a7796"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "-- TRAINING ---------------------------------------------------------------\n",
            "\n",
            "\n",
            " -- PREPROCESS SPATIAL TRANSCRIPTOMICS DATASET --------------------------------------------\n",
            "\n",
            "\n",
            "DATA (1/7): UC1_NI\n",
            "\n"
          ]
        },
        {
          "output_type": "error",
          "ename": "PathNotFoundError",
          "evalue": "nothing found at path ''",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mPathNotFoundError\u001b[0m                         Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-13-a6880ac2a3ca>\u001b[0m in \u001b[0;36m<cell line: 2>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# /!\\ Don't forget to import pre-trained Resnet50 \"pytorch_model.bin\" in model_directory_path=\"./resources\" from https://huggingface.co/timm/resnet50.tv_in1k/tree/main\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mtrain\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata_directory_path\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'./data'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmodel_directory_path\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"./resources\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
            "\u001b[0;32m<ipython-input-12-310f1cec02a1>\u001b[0m in \u001b[0;36mtrain\u001b[0;34m(data_directory_path, model_directory_path)\u001b[0m\n\u001b[1;32m     55\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     56\u001b[0m     \u001b[0;31m# Preprocess the spatial transcriptomics data for each sample in list_ST_name_data -> all preprocessed data saved in dir_processed_dataset\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 57\u001b[0;31m     preprocess_spatial_transcriptomics_data_train(list_ST_name_data, data_directory_path, dir_processed_dataset,\n\u001b[0m\u001b[1;32m     58\u001b[0m                                                   args.size_subset, args.target_patch_size, args.vis_width, args.show_extracted_images)\n\u001b[1;32m     59\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m<ipython-input-6-e372d9577db7>\u001b[0m in \u001b[0;36mpreprocess_spatial_transcriptomics_data_train\u001b[0;34m(list_ST_name_data, data_directory_path, dir_processed_dataset, size_subset, target_patch_size, vis_width, show_extracted_images)\u001b[0m\n\u001b[1;32m    161\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    162\u001b[0m         \u001b[0;31m# Load the spatial transcriptomics data from the .zarr format\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 163\u001b[0;31m         \u001b[0msdata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_zarr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata_directory_path\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34mf\"{name_data}.zarr\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    164\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    165\u001b[0m         \u001b[0;31m# Extract the list of gene names\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/spatialdata/_io/io_zarr.py\u001b[0m in \u001b[0;36mread_zarr\u001b[0;34m(store, selection)\u001b[0m\n\u001b[1;32m     54\u001b[0m     \u001b[0mA\u001b[0m \u001b[0mSpatialData\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     55\u001b[0m     \"\"\"\n\u001b[0;32m---> 56\u001b[0;31m     \u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mf_store_path\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_open_zarr_store\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstore\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     57\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     58\u001b[0m     \u001b[0mimages\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/spatialdata/_io/io_zarr.py\u001b[0m in \u001b[0;36m_open_zarr_store\u001b[0;34m(store)\u001b[0m\n\u001b[1;32m     29\u001b[0m     \u001b[0mA\u001b[0m \u001b[0mtuple\u001b[0m \u001b[0mof\u001b[0m \u001b[0mthe\u001b[0m \u001b[0mzarr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mGroup\u001b[0m \u001b[0mobject\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mthe\u001b[0m \u001b[0mpath\u001b[0m \u001b[0mto\u001b[0m \u001b[0mthe\u001b[0m \u001b[0mstore\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     30\u001b[0m     \"\"\"\n\u001b[0;32m---> 31\u001b[0;31m     \u001b[0mf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mstore\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstore\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mzarr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mGroup\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32melse\u001b[0m \u001b[0mzarr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstore\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmode\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"r\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     32\u001b[0m     \u001b[0;31m# workaround: .zmetadata is being written as zmetadata (https://github.com/zarr-developers/zarr-python/issues/1121)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     33\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstore\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstr\u001b[0m \u001b[0;34m|\u001b[0m \u001b[0mPath\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstore\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstartswith\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"http\"\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/zarr/convenience.py\u001b[0m in \u001b[0;36mopen\u001b[0;34m(store, mode, zarr_version, path, **kwargs)\u001b[0m\n\u001b[1;32m    135\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mopen_group\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_store\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmode\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mmode\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    136\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 137\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mPathNotFoundError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    138\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    139\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mPathNotFoundError\u001b[0m: nothing found at path ''"
          ]
        }
      ],
      "source": [
        "# /!\\ Don't forget to import pre-trained Resnet50 \"pytorch_model.bin\" in model_directory_path=\"./resources\" from https://huggingface.co/timm/resnet50.tv_in1k/tree/main\n",
        "train(data_directory_path='./data', model_directory_path=\"./resources\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "0OSDlFhPEx3P"
      },
      "outputs": [],
      "source": [
        "##########\n",
        "# functions for inference\n",
        "##########\n",
        "\n",
        "def embedding_and_load_data(name_data, dir_processed_dataset_test, test_embed_dir, args, device):\n",
        "    \"\"\"\n",
        "    Embedding of the images using the specified encoder and load the resulting data.\n",
        "\n",
        "    Args:\n",
        "    - name_data (str): The name of the data to process.\n",
        "    - dir_processed_dataset_test (str): Directory where the processed test dataset is stored.\n",
        "    - test_embed_dir (str): Directory where the embeddings should be saved.\n",
        "    - args (namespace): Arguments object containing parameters like encoder, batch_size, etc.\n",
        "    - device (torch.device): The device (CPU or GPU) to perform computations on.\n",
        "\n",
        "    Returns:\n",
        "    - assets (dict): Dictionary containing the 'barcodes', 'coords', and 'embeddings' from the embedded data.\n",
        "    \"\"\"\n",
        "\n",
        "    # Print the encoder being used\n",
        "    print(f\"Embedding images using {args.encoder} encoder\")\n",
        "\n",
        "    # Create encoder based on the specified model and load its weights\n",
        "    encoder = inf_encoder_factory(args.encoder)(args.weights_root)\n",
        "\n",
        "    # Define the path for the patches data (h5 file)\n",
        "    tile_h5_path = os.path.join(dir_processed_dataset_test, \"patches\", f'{name_data}.h5')\n",
        "\n",
        "    # Check if the file exists\n",
        "    assert os.path.isfile(tile_h5_path), f\"Patches h5 file not found at {tile_h5_path}\"\n",
        "\n",
        "    # Define the embedding output path\n",
        "    embed_path = os.path.join(test_embed_dir, f'{name_data}.h5')\n",
        "\n",
        "    # Generate the embeddings and save them to the defined path\n",
        "    generate_embeddings(embed_path, encoder, device, tile_h5_path, args.batch_size, args.num_workers, overwrite=args.overwrite)\n",
        "\n",
        "    # Load the embeddings and related assets\n",
        "    assets, _ = read_assets_from_h5(embed_path)\n",
        "\n",
        "    # Extract cell IDs and convert to a list of strings\n",
        "    # The cell IDs are not necessary because the images are kept in the same order as the gene expression data\n",
        "    cell_ids = assets['barcodes'].flatten().astype(str).tolist()\n",
        "\n",
        "    return assets\n",
        "\n",
        "\n",
        "def load_models_from_directories(base_path):\n",
        "    \"\"\"\n",
        "    Load all trained regression models (one model for each cross-validation split)\n",
        "    Load 'model.pkl' from each directory within the base_path.\n",
        "\n",
        "    :param base_path: The parent directory containing split subdirectories.\n",
        "    :return: A dictionary where keys are directory names and values are the loaded models.\n",
        "    \"\"\"\n",
        "\n",
        "    models = {}\n",
        "    for name in os.listdir(base_path):\n",
        "        dir_path = os.path.join(base_path, name)\n",
        "        if os.path.isdir(dir_path):\n",
        "            model_path = os.path.join(dir_path, 'model.pkl')\n",
        "            if os.path.exists(model_path):\n",
        "                models[name] = joblib.load(model_path)\n",
        "                print(f\"Loaded model from {model_path}\")\n",
        "            else:\n",
        "                print(f\"'model.pkl' not found in {dir_path}\")\n",
        "\n",
        "    return models\n",
        "\n",
        "\n",
        "def predict_and_aggregate_models(X_test, results_dir):\n",
        "    \"\"\"\n",
        "    Load models from the given directory, make predictions on the test set (X_test),\n",
        "    aggregate the predictions by averaging, and set negative predictions to 0.\n",
        "\n",
        "    Args:\n",
        "    - X_test (np.array): The test data to make predictions on.\n",
        "    - results_dir (str): Directory containing the saved models.\n",
        "\n",
        "    Returns:\n",
        "    - np.array: The aggregated predictions.\n",
        "    \"\"\"\n",
        "\n",
        "    # Load models from the specified directory\n",
        "    models = load_models_from_directories(results_dir)\n",
        "\n",
        "    # Initialize a list to store predictions\n",
        "    predictions = []\n",
        "\n",
        "    # Iterate through each model and make predictions\n",
        "    for split_name in models.keys():\n",
        "        preds = models[split_name].predict(X_test)\n",
        "        predictions.append(preds)\n",
        "\n",
        "    # Stack the predictions into a 2D array (models x samples)\n",
        "    predictions = np.stack(predictions)\n",
        "\n",
        "    # Aggregate predictions by calculating the mean across all models\n",
        "    average_predictions = np.mean(predictions, axis=0)\n",
        "\n",
        "    # Set any negative predictions to 0\n",
        "    average_predictions = np.where(average_predictions < 0, 0.0, average_predictions)\n",
        "\n",
        "    del models\n",
        "\n",
        "    return average_predictions"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "W6R_RAbOtxHh"
      },
      "source": [
        "## Infer function\n",
        "\n",
        "The `infer` function loads trained models and performs inference on a new dataset.\n",
        "\n",
        "1. Prepare the necessary directories and load the configuration parameters from the previously trained model.\n",
        "2. The test data, provided as a Zarr file, is read and specific subsets of the data (test and validation groups) are extracted.\n",
        "3. Preprocess the data into image patches (X_test).\n",
        "4. Generate embeddings for the test data and applies the trained models for regression predictions.\n",
        "5. Format the predictions for submission."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "40k_Y0a--9up"
      },
      "outputs": [],
      "source": [
        "# In the inference function, the trained model is loaded and used to make inferences on a\n",
        "# sample of data that matches the characteristics of the training test.\n",
        "def infer(\n",
        "    data_file_path: str,  # Path to a test dataset (in Zarr format) to perform inference on.\n",
        "    model_directory_path: str  # Path to save the trained model and results\n",
        "):\n",
        "    ### Prepare Directories ###\n",
        "\n",
        "    # Extract the name of the dataset from the file path (without extension)\n",
        "    name_data = os.path.splitext(os.path.basename(data_file_path))[0]\n",
        "    print(f\"\\n-- {name_data} INFERENCE ---------------------------------------------------------------\\n\")\n",
        "    print(data_file_path)\n",
        "\n",
        "    # Previous directory where models and results are stored\n",
        "    dir_models_and_results = os.path.join(model_directory_path, f\"ST_pred_results\")\n",
        "    # Load training configuration parameters\n",
        "    config_path = os.path.join(dir_models_and_results, \"config.json\")\n",
        "    with open(config_path, 'r') as f:\n",
        "        args_dict = json.load(f)\n",
        "    args = SimpleNamespace(**args_dict)\n",
        "\n",
        "    # Directory for processed test dataset (temporary storage)\n",
        "    dir_processed_dataset_test = os.path.join(\"/tmp\", f\"processed_dataset_test\")\n",
        "    os.makedirs(dir_processed_dataset_test, exist_ok=True)\n",
        "\n",
        "    # Directory to store the test data embeddings (temporary storage)\n",
        "    test_embed_dir = os.path.join(dir_processed_dataset_test, \"ST_data_emb\")\n",
        "    os.makedirs(test_embed_dir, exist_ok=True)\n",
        "\n",
        "    # Set device to GPU if available, else use CPU!!\n",
        "    device = torch.device(\"cuda\" if torch.cuda.is_available() else \"cpu\")\n",
        "\n",
        "    ### Preprocess and Embedding Data + Regression inference ###\n",
        "\n",
        "    # Read the spatial data from the provided file\n",
        "    sdata = sd.read_zarr(data_file_path)\n",
        "\n",
        "    # Extract cell IDs for test and validation groups\n",
        "    cell_ids = list(sdata[\"cell_id-group\"].obs.query(\"group == 'test' or group == 'validation'\")[\"cell_id\"])\n",
        "\n",
        "    # Extract gene names from the spatial data\n",
        "    gene_names = list(sdata[\"anucleus\"].var.index)\n",
        "\n",
        "    # Preprocess the test data for embedding (patch extraction)\n",
        "    preprocess_spatial_transcriptomics_data_test(name_data, sdata, cell_ids, dir_processed_dataset_test,\n",
        "                                                 args.target_patch_size, args.vis_width, args.show_extracted_images)\n",
        "\n",
        "    print(f\"\\n-- {name_data} EMBEDDING--\\n\")\n",
        "    # Generate and load the embeddings for the test data\n",
        "    assets = embedding_and_load_data(name_data, dir_processed_dataset_test, test_embed_dir, args, device)\n",
        "\n",
        "    # Extract embeddings features for prediction\n",
        "    X_test = assets[\"embeddings\"]\n",
        "    print(\"Embedding shape (X_test):\", X_test.shape)\n",
        "\n",
        "    print(f\"\\n-- {name_data} REGRESSION PREDICTIONS--\\n\")\n",
        "    # Make predictions and aggregate results across cross-validation regression models\n",
        "    average_predictions = predict_and_aggregate_models(X_test, args.results_dir)\n",
        "\n",
        "    ### Prepare and Return Predictions ###\n",
        "\n",
        "    # Convert the predictions to a DataFrame (the gene expression value must be rounded to two decimal places)\n",
        "    prediction = pd.DataFrame(np.round(average_predictions, 2), index=cell_ids, columns=gene_names)\n",
        "    # Reset index to have 'cell_id' as a column\n",
        "    prediction = prediction.reset_index(names=\"cell_id\")\n",
        "\n",
        "    # Melt the DataFrame to the expected output for the challenge\n",
        "    prediction = prediction.melt(id_vars=\"cell_id\", var_name=\"gene\", value_name=\"prediction\")\n",
        "    # prediction = prediction.sort_values(by=[\"cell_id\", \"gene\"]).reset_index(drop=True)\n",
        "\n",
        "    # Free memory by deleting large variables and performing garbage collection\n",
        "    del average_predictions, sdata, X_test, assets\n",
        "    gc.collect()\n",
        "\n",
        "    print(f\"\\n-- {name_data} PREDICTION DONE\\n\")\n",
        "\n",
        "    # Return the final prediction DataFrame\n",
        "    return prediction"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "v2T41EUZgau3"
      },
      "outputs": [],
      "source": [
        "prediction = infer(\n",
        "    data_file_path=\"./data/UC1_NI.zarr\",\n",
        "    model_directory_path=\"./resources\"\n",
        ")\n",
        "prediction"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "a5w-KSO52hDj"
      },
      "outputs": [],
      "source": [
        "# This command is running a local test with your submission\n",
        "# making sure that your submission can be accepted by the system\n",
        "# crunch.test(\n",
        "#     no_determinism_check=True,\n",
        "# )"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "47HkOHSSynG5"
      },
      "source": [
        "**You can download this notebook and then submit it at https://hub.crunchdao.com/competitions/broad-1/submit/**"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "kppG1inFIopT"
      },
      "source": [
        "Steps to submit the notebook:\n",
        "\n",
        "- Download locally the pre-trained model Resnet50 `pytorch_model.bin`: https://huggingface.co/timm/resnet50.tv_in1k/tree/main (because internet is restricted on CrunchDAO submission platform)\n",
        "- Create a directory named `model`\n",
        "- Place the file `pytorch_model.bin` inside the newly created `model` directory\n",
        "- Submit the `model` directory in the \"Model Files\" field\n",
        "- Run notebook in the cloud with CPU environnment\n",
        "- Set Training to Yes\n",
        "\n",
        "![notebook_submit_steps](https://raw.githubusercontent.com/crunchdao/competitions/refs/heads/master/competitions/broad-1/quickstarters/resnet50-plus-ridge/images/notebook_submit_steps.png)"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "machine_shape": "hm",
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.10.11"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}